{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XZZ\n"
     ]
    }
   ],
   "source": [
    "'''CODE CHALLENGE: Solve the Peptide Identification Problem.'''\n",
    "    \n",
    "def PeptideIdentification(Spectral_v,Proteome,prot):\n",
    "\n",
    "    '''\n",
    "    Find a peptide from a proteome with maximum score against a spectrum\n",
    "    !!! Add [0] to Spectral Vector\n",
    "    Spectral Vector= [0,0,0,4,-2,-3,-1,-7,6,5,3,2,1,9,3,-8,0,3,1,2,1,8] #A space-delimited spectral vector Spectrum'\n",
    "    INPUT:\n",
    "    Spectral_v=[0,0,0,0,4,-2,-3,-1,-7,6,5,3,2,1,9,3,-8,0,3,1,2,1,8]\n",
    "    Proteome=XZZXZXXXZXZZXZXXZ # An amino acid string Proteome.\n",
    "    prot={'X':4,'Z':5} # aa <-> mass dictionary\n",
    "    OUTPUT= ZXZXX # A substring of Proteome with maximum score against Spectrum'.\n",
    "    '''    \n",
    "    \n",
    "    def ScorePeptideAgainstSpectralVector(Peptide,Spectral_v,prot):\n",
    "        index=0\n",
    "        score=0\n",
    "        max_pep_weight=len(Spectral_v)-1\n",
    "        for n in range(len(Peptide)):\n",
    "            aa=Peptide[n]\n",
    "            index+=prot[aa]\n",
    "            if index>max_pep_weight:\n",
    "                return 'Peptide too massive'\n",
    "            else:\n",
    "                score+=Spectral_v[index]\n",
    "        if index<max_pep_weight:\n",
    "             return 'Peptide too light'\n",
    "            \n",
    "        return score\n",
    "\n",
    "    #############################################################################\n",
    "    pep_score=[]\n",
    "    max_score=-float('Inf')\n",
    "    max_pep=''\n",
    "    max_pep_len=len(Spectral_v)-1\n",
    "    for n in range(0,len(Proteome)):\n",
    "        for i in range(n+1,len(Proteome)):\n",
    "            peptide= Proteome[n:i]\n",
    "            if peptide in pep_score:\n",
    "                pass\n",
    "            else:\n",
    "                pep_score+=[peptide]\n",
    "                score=ScorePeptideAgainstSpectralVector(peptide,Spectral_v,prot)\n",
    "                if score == 'Peptide too massive':\n",
    "                    break\n",
    "                if score == 'Peptide too light':\n",
    "                    pass\n",
    "                elif score>max_score:\n",
    "                    max_score=score\n",
    "                    max_pep=peptide\n",
    "    return max_pep\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "#spec=[0,0,0,4,-2,-3,-1,-7,6,5,3,2,1,9,3,-8,0,3,1,2,1,8]\n",
    "spec=[4, -3, -2, 3, 3, -4, 5, -3, -1, -1, 3, 4, 1, 3]\n",
    "Proteome='XZZXZXXXZXZZXZXXZ'\n",
    "prot={'X':4,'Z':5}\n",
    "\n",
    "'''\n",
    "f=open('input.txt','r')\n",
    "#f=open('../../Downloads/dataset_11866_2(4).txt','r')\n",
    "#f=open('../../Downloads/rosalind_ba11f.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "spec=map(int,lines[0].split())\n",
    "Proteome=lines[1]\n",
    "#print Proteome\n",
    "#print spec\n",
    "#prot={'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'H': 137, 'M': 131, 'I':113, 'L': 113, 'N': 114, 'K':128, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163}\n",
    "'''\n",
    "\n",
    "spec=[0]+spec\n",
    "#print spec\n",
    "#print ScorePeptideAgainstSpectralVector('VEAAFYAKITNHSGH',spec,prot)\n",
    "print PeptideIdentification(spec,Proteome,prot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''PSMSearch(SpectralVectors, Proteome, threshold).\n",
    "    PSMSet ← an empty set\n",
    "    for each vector Spectrum' in SpectralVectors\n",
    "          Peptide ← PeptideIdentification(Spectrum', Proteome)\n",
    "          if Score(Peptide, Spectrum) ≥ threshold\n",
    "              add the PSM (Peptide, Spectrum') to PSMSet\n",
    "    return PSMSet'''\n",
    "def PSMsearch(SpectralVectors, Proteome, threshold,prot=[]):\n",
    "    \n",
    "    ''' \n",
    "    Identify Peptide-Spectrum Matches by matching spectra against a proteome.\n",
    "    INPUT:\n",
    "    SpectralVectors = [[-1, 5, -4, 5, 3, -1, -4, 5, -1, 0, 0, 4, -1, 0, 1, 4, 4, 4], [-4, 2, -2, -4, 4, -5, -1, 4, -1, 2, 5, -3, -1, 3, 2, -3]] # A list of spectral vectors \n",
    "    Proteome = XXXZXZXXZXZXXXZXXZX # an amino acid string Proteome\n",
    "    threshold = 5 # a integer score threshold\n",
    "    prot = {'X':4,'Z':5} # aa <-> mass dictionary\n",
    "    OUTPUT = ['XZXZ'] # A list of  all unique Peptide-Spectrum Matches scoring at least as high threshold.\n",
    "    '''\n",
    " \n",
    "    def ScorePeptideAgainstSpectralVector(Peptide,Spectral_v,prot):\n",
    "        index=0\n",
    "        score=0\n",
    "        max_pep_weight=len(Spectral_v)-1\n",
    "        for n in range(len(Peptide)):\n",
    "            aa=Peptide[n]\n",
    "            index+=prot[aa]\n",
    "            if index>max_pep_weight:\n",
    "                return 'Peptide too massive'\n",
    "            else:\n",
    "                score+=Spectral_v[index]\n",
    "        if index<max_pep_weight:\n",
    "             return 'Peptide too light'\n",
    "        return score\n",
    "    \n",
    "    def PeptideIdentification2(Spectral_v,Proteome,prot):\n",
    "        '''\n",
    "        Find a peptide from a proteome with maximum score against a spectrum. To run faster, this version first evaluate the min size of peptide and \n",
    "        skip peptides too small\n",
    "        !!! Add [0] to Spectral Vector\n",
    "        Spectral Vector= [0,0,0,4,-2,-3,-1,-7,6,5,3,2,1,9,3,-8,0,3,1,2,1,8] #A list of spectral vector Spectrum'\n",
    "        INPUT:\n",
    "        Spectral_v=[0,0,0,0,4,-2,-3,-1,-7,6,5,3,2,1,9,3,-8,0,3,1,2,1,8]\n",
    "        Proteome=XZZXZXXXZXZZXZXXZ # An amino acid string Proteome.\n",
    "        prot={'X':4,'Z':5} # aa <-> mass dictionary\n",
    "        OUTPUT= ZXZXX # A substring of Proteome with maximum score against Spectrum'.\n",
    "        '''    \n",
    "        pep_score=[]\n",
    "        max_score=-float('Inf')\n",
    "        max_pep=''\n",
    "        max_pep_weight=len(Spectral_v)-1\n",
    "        max_aa=max(prot.values())\n",
    "        if max_pep_weight%max_aa>0:\n",
    "            min_pep_len=max_pep_weight/max_aa+1\n",
    "        else:\n",
    "            min_pep_len=max_pep_weight/max_aa\n",
    "        for n in range(0,len(Proteome)):\n",
    "            for i in range(n+min_pep_len,len(Proteome)):\n",
    "                peptide= Proteome[n:i]\n",
    "                if peptide in pep_score:\n",
    "                    pass\n",
    "                else:\n",
    "                    pep_score+=[peptide]\n",
    "                    score=ScorePeptideAgainstSpectralVector(peptide,Spectral_v,prot)\n",
    "                    if score == 'Peptide too massive':\n",
    "                        break\n",
    "                    if score == 'Peptide too light':\n",
    "                        pass\n",
    "                    elif score>max_score:\n",
    "                        max_score=score\n",
    "                        max_pep=peptide\n",
    "        return max_pep,max_score\n",
    "    ###############################################################################\n",
    "    prot=prot\n",
    "    PSMset=[]\n",
    "    for Spectrum_v in SpectralVectors:\n",
    "        Spectrum_v=[0]+Spectrum_v #prepond spectral vector with 0\n",
    "        Peptide,score = PeptideIdentification2(Spectrum_v,Proteome,prot)\n",
    "        if score>=threshold:\n",
    "            PSMset+=[Peptide]\n",
    "    return PSMset\n",
    "        \n",
    "############\n",
    "\n",
    "f=open('input.txt','r')\n",
    "#f=open('input2.txt','r')\n",
    "#f=open('../../Downloads/dataset_11866_5.txt','r')\n",
    "#f=open('../../Downloads/rosalind_ba11g.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "SpectralVectors=[]\n",
    "for n in range(len(lines)):\n",
    "    line=lines[n]\n",
    "    try:\n",
    "        if len(map(int,line.split()))>1:\n",
    "            SpectralVectors+=[map(int,line.split())]\n",
    "        else:\n",
    "            threshold=int(line)\n",
    "    except ValueError:\n",
    "        Proteome=line\n",
    "    \n",
    "print SpectralVectors\n",
    "print Proteome\n",
    "print threshold\n",
    "#prot={'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'H': 137, 'M': 131, 'I':113, 'L': 113, 'N': 114, 'K':128, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163}\n",
    "prot={'X':4,'Z':5}\n",
    "\n",
    "\n",
    "        \n",
    "print '\\n'.join(PSMsearch(SpectralVectors, Proteome, threshold,prot=prot))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"CODE CHALLENGE: Solve the Size of Spectral Dictionary Problem.\\n     Given: A spectral vector Spectrum', an integer threshold, and an integer max_score.\\n     Return: The size of the dictionary Dictionarythreshold(Spectrum').\\n\\n\""
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''CODE CHALLENGE: Solve the Size of Spectral Dictionary Problem.\n",
    "     Given: A spectral vector Spectrum', an integer threshold, and an integer max_score.\n",
    "     Return: The size of the dictionary Dictionarythreshold(Spectrum').\n",
    "\n",
    "'''\n",
    "import time\n",
    "start=time.time()\n",
    "\n",
    "#f=open('input.txt','r')\n",
    "#f=open('input2.txt','r')\n",
    "#f=open('../../Downloads/dataset_11866_8.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba11h.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "Spectrum=[0]+map(int,lines[0].split())\n",
    "#print Spectrum\n",
    "threshold=int(lines[1])\n",
    "max_score=int(lines[2])\n",
    "#print len(Spectrum)-1\n",
    "#print threshold\n",
    "#print max_score\n",
    "prot={'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'H': 137, 'M': 131, 'I':113, 'L': 113, 'N': 114, 'K':128, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163}\n",
    "#prot={'X':4,'Z':5}\n",
    "\n",
    "def DicThresholdSpec(Spectrum, threshold, max_score,prot): \n",
    "    '''\n",
    "    Find the number of peptides that match Spectrum with a score between threshold and max_score.\n",
    "    Uses 'prot' aa masse table. \n",
    "    INPUT:\n",
    "    Spectrum=[0, 4, -3, -2, 3, 3, -4, 5, -3, -1, -1, 3, 4, 1, 3]\n",
    "    threshold=1\n",
    "    max_score=8\n",
    "    prot={'X':4,'Z':5}\n",
    "    OUTPUT = 3\n",
    "    '''\n",
    "    import numpy as np\n",
    "    m=len(Spectrum)-1\n",
    "    size = np.zeros((m+1,max_score+1),dtype=int)\n",
    "    size[0,0]=1\n",
    "    for i in range(1,m+1):\n",
    "        for t in range(max_score+1): \n",
    "            size[i,t] = sum([size[i-aa_weight,t-Spectrum[i]] for aa_weight in prot.values() if (i-aa_weight)>=0 and (t-Spectrum[i])>=0 and (t-Spectrum[i])<=max_score])\n",
    "    result=sum(size[m,t] for t in range(threshold,max_score+1))        \n",
    "    return result\n",
    "\n",
    "\n",
    "size= DicThresholdSpec(Spectrum, threshold, max_score,prot)\n",
    "print size\n",
    "print 'Time it took',time.time()-start,'s'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.284375e-05\n",
      "Time it took 1.51231789589 s\n"
     ]
    }
   ],
   "source": [
    "'''CODE CHALLENGE: Solve the Probability of Spectral Dictionary Problem.'''\n",
    "\n",
    "import time\n",
    "start=time.time()\n",
    "\n",
    "#f=open('input.txt','r')\n",
    "#f=open('input2.txt','r')\n",
    "#f=open('../../Downloads/dataset_11866_11.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba11i.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "Spectrum=[0]+map(int,lines[0].split())\n",
    "#print Spectrum\n",
    "threshold=int(lines[1])\n",
    "max_score=int(lines[2])\n",
    "#print len(Spectrum)-1\n",
    "#print threshold\n",
    "#print max_score\n",
    "prot={'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'H': 137, 'M': 131, 'I':113, 'L': 113, 'N': 114, 'K':128, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163}\n",
    "#prot={'X':4,'Z':5}\n",
    "\n",
    "def Pr_DicThresholdSpec(Spectrum, threshold, max_score,prot): \n",
    "    '''\n",
    "    Find the probability for a peptide match against Spectrum with a score between threshold and max_score.\n",
    "    Find the probability of the spectral dictionary for a given spectrum and score threshold.\n",
    "    Uses 'prot' aa masse table. \n",
    "    INPUT:\n",
    "    Spectrum=[0, 4, -3, -2, 3, 3, -4, 5, -3, -1, -1, 3, 4, 1, 3]\n",
    "    threshold=1\n",
    "    max_score=8\n",
    "    prot={'X':4,'Z':5}\n",
    "    OUTPUT = 0.375\n",
    "    '''\n",
    "    import numpy as np\n",
    "    m=len(Spectrum)-1\n",
    "    size = np.zeros((m+1,max_score+1),dtype=float)\n",
    "    size[0,0]=1\n",
    "    dimensions=float(1)/len(prot) # can alos be written 1./len(prot)\n",
    "    for i in range(1,m+1):\n",
    "        for t in range(max_score+1): \n",
    "            size[i,t] = dimensions * sum([size[i-aa_weight,t-Spectrum[i]] for aa_weight in prot.values() if (i-aa_weight)>=0 and (t-Spectrum[i])>=0 and (t-Spectrum[i])<=max_score])\n",
    "    result=sum(size[m,t] for t in range(threshold,max_score+1))        \n",
    "    return result\n",
    "\n",
    "\n",
    "size= Pr_DicThresholdSpec(Spectrum, threshold, max_score,prot)\n",
    "print size\n",
    "print 'Time it took',time.time()-start,'s'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XXZ\n",
      "[0, 4, 8, 13]\n",
      "15\n",
      "XX(-1)Z(+2)\n",
      "the script took  0.101531028748  s.\n"
     ]
    }
   ],
   "source": [
    "'''CODE CHALLENGE 5: Solve the Spectral Alignment Problem.'''\n",
    "import numpy as np\n",
    "import time\n",
    "import pandas as pd\n",
    "\n",
    "def PeptideToVector2(peptide,prot=[]):\n",
    "    \n",
    "    '''\n",
    "    Convert a peptide into a binary peptide vector.\n",
    "    INPUT = XZZXX # A peptide P.\n",
    "    OUTPUT = 0001000010000100010001 # The peptide vector of P.\n",
    "    prot={'X':4,'Z':5}\n",
    "    '''\n",
    "    prot=prot\n",
    "    vector_len=sum([prot[n] for n in peptide])\n",
    "    vector=[0]\n",
    "    for n in range(1,len(peptide)+1):\n",
    "        mass=sum([prot[n] for n in peptide[:n]])\n",
    "        vector+=[mass]\n",
    "    return vector\n",
    "\n",
    "\n",
    "Peptide='XXZ'\n",
    "Spectrum=[0, 4, -3, -2, 3, 3, -4, 5, -3, -1, -1, 3, 4, 1, -1]\n",
    "k=2\n",
    "\n",
    "\n",
    "prot={'X':4,'Z':5}\n",
    "#prot={'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'H': 137, 'M': 131, 'I':113, 'L': 113, 'N': 114, 'K':128, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163}\n",
    "\n",
    "'''\n",
    "#f=open('input.txt','r')\n",
    "#f=open('input2.txt','r')\n",
    "#f=open('../../Downloads/dataset_11866_14.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba11j.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "Peptide=lines[0]\n",
    "Spectrum=[0]+map(int,lines[1].split(' '))\n",
    "k=int(lines[2])\n",
    "'''\n",
    "start=time.time()\n",
    "\n",
    "print Peptide\n",
    "print PeptideToVector2(Peptide,prot=prot)\n",
    "print len(Spectrum)\n",
    "\n",
    "def SpectralAlignment(Peptide,Spectrum,k,prot):\n",
    "    '''\n",
    "    Given a peptide and a spectral vector, find a modified variant of this peptide that maximizes the peptide-spectrum score\n",
    "    among all variants of the peptides with up to k modifications.\n",
    "    \n",
    "    INPUT:\n",
    "    Peptide='XXZ'\n",
    "    Spectrum=[0, 4, -3, -2, 3, 3, -4, 5, -3, -1, -1, 3, 4, 1, -1]\n",
    "    k=2\n",
    "    prot={'X':4,'Z':5}\n",
    "    \n",
    "    OUTPUT= \n",
    "    \n",
    "    '''\n",
    "\n",
    "    ####### create dictionnary of dataframes and diff dictionnary ############\n",
    "    \n",
    "    pep_vector=[0]\n",
    "    for n in range(1,len(Peptide)+1):\n",
    "        mass=sum([prot[n] for n in Peptide[:n]])\n",
    "        pep_vector+=[mass]\n",
    "        \n",
    "    \n",
    "    df=pd.DataFrame(np.zeros((len(pep_vector),len(Spectrum)),dtype=float))\n",
    "    df.index=pep_vector\n",
    "    dfdic={}\n",
    "    for n in range(k+1):\n",
    "        dfdic[n]=df.copy()\n",
    "\n",
    "    diff={}\n",
    "    for n in range(1,len(pep_vector)):\n",
    "        index=pep_vector[n]\n",
    "        diff[index]=pep_vector[n]-pep_vector[n-1]\n",
    "\n",
    "    ###### initiate tables and backtrack ############\n",
    "\n",
    "    backtrack={}\n",
    "    backtrack['[0, 0, 0]']=[[0,0,0]]\n",
    "    for t in range(k+1):\n",
    "        for i in pep_vector:\n",
    "            for j in range(len(Spectrum)):\n",
    "                dfdic[t].loc[i,j]=-float('Inf')\n",
    "                if j==0 or i==0:\n",
    "                    coord=[t,i,j]\n",
    "        dfdic[t].loc[0,0]=0    \n",
    "\n",
    "    ######### fill in first table t=0 #######\n",
    "    t=0\n",
    "    for i in pep_vector[1:]:\n",
    "        for j in [vec for vec in pep_vector[1:] if vec in range(len(Spectrum))]:\n",
    "            coord=[t,i,j]\n",
    "            #print coord\n",
    "\n",
    "            if j-diff[i]>=0:\n",
    "\n",
    "                score_tpred=dfdic[t].loc[i-diff[i],j-diff[i]]\n",
    "                dfdic[t].loc[i,j]=Spectrum[j]+score_tpred\n",
    "\n",
    "                predec=[t,i-diff[i],j-diff[i]]\n",
    "\n",
    "                try:\n",
    "                    backtrack[str(coord)]=backtrack[str(predec)]+[coord]\n",
    "                except KeyError:\n",
    "                    pass\n",
    "\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "    ####### fill-in tables t>0 #############\n",
    "\n",
    "    for t in range(1,k+1):\n",
    "        for i in pep_vector[1:]:\n",
    "            for j in range(1,len(Spectrum)):\n",
    "                coord=[t,i,j]\n",
    "                idx_pred=[]\n",
    "                if j-diff[i]<0:\n",
    "                    score_tpred=-float('Inf')\n",
    "                else:    \n",
    "                    score_tpred=dfdic[t].loc[i-diff[i],j-diff[i]]\n",
    "                score_opred=list(dfdic[t-1].loc[i-diff[i]][:j])\n",
    "                all_scores=[score_tpred]+score_opred\n",
    "\n",
    "                score=max(all_scores)\n",
    "                if score !=-float('Inf'):\n",
    "                    dfdic[t].loc[i,j]=Spectrum[j]+score\n",
    "\n",
    "                    # in case of multiple max_scores, discard the ones that don't lead to source (0,0,0)\n",
    "                    max_scores=[a for a, b in enumerate(all_scores) if b == score]\n",
    "                    if len(max_scores)>1: \n",
    "                        for n in max_scores:\n",
    "                            if n==0:\n",
    "                                predec=[t,i-diff[i],j-diff[i]]\n",
    "                                try:\n",
    "                                    backtrack[str(predec)]\n",
    "                                    break\n",
    "                                except KeyError:\n",
    "                                    pass\n",
    "                            else:\n",
    "                                predec=[t-1,i-diff[i],n-1]\n",
    "                                try:\n",
    "                                    backtrack[str(predec)]\n",
    "                                    break\n",
    "                                except KeyError:\n",
    "                                    pass\n",
    "                    else:\n",
    "                        if score == score_tpred:\n",
    "                            predec=[t,i-diff[i],j-diff[i]]\n",
    "                        else:\n",
    "                            predec=[t-1,i-diff[i],score_opred.index(max(score_opred))]\n",
    "\n",
    "                try:\n",
    "                    backtrack[str(coord)]=backtrack[str(predec)]+[coord]\n",
    "                except KeyError:\n",
    "                    pass\n",
    "\n",
    "    ####### find max value and format result ############\n",
    "\n",
    "    max_score=max([dfdic[t].loc[pep_vector[-1],len(Spectrum)-1] for t in range(k+1)])\n",
    "    t_idx_max=[t for t in range(k+1) if dfdic[t].loc[pep_vector[-1],len(Spectrum)-1]==max_score]\n",
    "    if len(t_idx_max)==1:\n",
    "        coord_max=[t_idx_max[0],pep_vector[-1],len(Spectrum)-1]\n",
    "    else:\n",
    "        for item in t_idx_max:\n",
    "            coord_max=[t_idx_max,pep_vector[-1],len(Spectrum)-1]\n",
    "            try:\n",
    "                backtrack[str(coord_max)]\n",
    "            except KeyError:\n",
    "                pass\n",
    "\n",
    "    modif_pep_vector=[j for t,i,j in backtrack[str(coord_max)]]\n",
    "\n",
    "    result=''\n",
    "    for n in range(1,len(pep_vector)):\n",
    "        aa=Peptide[n-1]    \n",
    "        delta_weight=modif_pep_vector[n]-modif_pep_vector[n-1]-prot[aa]\n",
    "        if delta_weight==0:\n",
    "            result+=aa\n",
    "        else:\n",
    "            result+=aa+'('+'%+d' % delta_weight+')'\n",
    "    return result\n",
    "\n",
    "print SpectralAlignment(Peptide,Spectrum,k,prot)\n",
    "\n",
    "print 'the script took ',time.time()-start,' s.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{0:     0   1   2   3   4   5   6   7   8   9   10  11  12  13  14\n",
      "0    0  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA\n",
      "4   NA  NA  NA  NA   3  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA\n",
      "8   NA  NA  NA  NA  NA  NA  NA  NA   0  NA  NA  NA  NA  NA  NA\n",
      "13  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA   1  NA, 1:     0   1   2   3   4   5   6   7   8   9   10  11  12  13  14\n",
      "0   NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA\n",
      "4   NA   4  -3  -2  NA   3  -4   5  -3  -1  -1   3   4   1  -1\n",
      "8   NA  NA  NA  NA  NA   7  -1   8  NA   2   2   8   7   4   2\n",
      "13  NA  NA  NA  NA  NA  NA  NA  NA  NA  -1   6   3  12  NA   1, 2:     0   1   2   3   4   5   6   7   8   9   10  11  12  13  14\n",
      "0   NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA\n",
      "4   NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA  NA\n",
      "8   NA  NA   1   2   7  NA   0   9   2   4   4  NA   9   6   4\n",
      "13  NA  NA  NA  NA  NA  NA   3  12   5   7   7  11  13   9   7}\n"
     ]
    }
   ],
   "source": [
    "dd=dfdic.copy()\n",
    "#print dd\n",
    "\n",
    "for t in range(3):\n",
    "    for i in pep_vector:\n",
    "        for j in range(len(Spectrum)):\n",
    "            try:\n",
    "                backtrack[str([t,i,j])]\n",
    "            except KeyError:\n",
    "                dd[t].loc[i,j]='NA'\n",
    "print dd\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 642,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[2, 619, 420]]\n",
      "[131, 260, 389, 504, 619, 756]\n",
      "[131, 260, 389, 504, 619, 756]\n",
      "[131, 260, 389, 504, 619, 756]\n",
      "1 3 3 1\n"
     ]
    }
   ],
   "source": [
    "print backtrack[str([2, 619, 420])]\n",
    "\n",
    "print pep_vector[1:]\n",
    "print [vec for vec in pep_vector[1:] if vec in range(len(Spectrum))]\n",
    "print sorted(set(pep_vector[1:]) & set(range(len(Spectrum))))\n",
    "\n",
    "a=1\n",
    "b=max([2,3])\n",
    "print a,b,max(a,b),[a,b].index(max(a,b))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          0         1         2         3         4         5         6   \\\n",
      "0   0.000000      -inf      -inf      -inf      -inf      -inf      -inf   \n",
      "4       -inf  4.000000 -3.000000 -2.000000  3.000000  3.000000 -4.000000   \n",
      "8       -inf      -inf      -inf      -inf      -inf  7.000000 -1.000000   \n",
      "13      -inf      -inf      -inf      -inf      -inf      -inf      -inf   \n",
      "\n",
      "          7         8         9         10        11         12        13  \\\n",
      "0       -inf      -inf      -inf      -inf      -inf       -inf      -inf   \n",
      "4   5.000000 -3.000000 -1.000000 -1.000000  3.000000   4.000000  1.000000   \n",
      "8   8.000000  0.000000  2.000000  2.000000  8.000000   7.000000  4.000000   \n",
      "13      -inf      -inf -1.000000  6.000000  3.000000  12.000000  1.000000   \n",
      "\n",
      "          14  \n",
      "0       -inf  \n",
      "4  -1.000000  \n",
      "8   2.000000  \n",
      "13  1.000000  \n",
      "0        -inf\n",
      "1    4.000000\n",
      "2   -3.000000\n",
      "Name: 4, dtype: float64\n",
      "4.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 607,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print dfdic[1]\n",
    "print dfdic[1].loc[4][:3]\n",
    "DataFrame=dfdic[1].loc[4][:3]\n",
    "print max(list(dfdic[1].loc[4][:3]))\n",
    "DataFrame.idxmax()\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 575,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0, 0, 0]]\n",
      "[('a', 1), ('c', 3), ('b', 2)]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('a', 1), ('c', 3), ('b', 2)]"
      ]
     },
     "execution_count": 575,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "items=[[1,2,3],[4,5,6],[7,8,9]]\n",
    "idx_pred=[[0, 0, 0]]\n",
    "#print [[t,i,j] for t,i,j in items]\n",
    "print [[t,i,j] for t,i,j in idx_pred]\n",
    "pep_vector[1:]\n",
    "\n",
    "dic={'a':1,'b':2,'c':3}\n",
    "print [(k,v) for k,v in dic.items()]\n",
    "dic.items()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unexpected indent (<ipython-input-6-f3dfbf042efd>, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-6-f3dfbf042efd>\"\u001b[0;36m, line \u001b[0;32m2\u001b[0m\n\u001b[0;31m    '''\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m unexpected indent\n"
     ]
    }
   ],
   "source": [
    "\n",
    "            '''\n",
    "            if j-diff[i]<0:\n",
    "                score_tpred=-float('Inf')\n",
    "            else:\n",
    "                score_tpred=dfdic[t].loc[i-diff[i],j-diff[i]]\n",
    "           \n",
    "            else:\n",
    "                score_opred=[dfdic[t-1].loc[i-diff[i],jn] for jn in range(j)]\n",
    "            score=Spectrum[j]+max([score_tpred]+score_opred)\n",
    "            #print 'predec_scores',[score_tpred]+score_opred\n",
    "            dfdic[t].loc[i,j]+=score\n",
    "            if t==0:\n",
    "                if j-diff[i]<0:\n",
    "                    break\n",
    "                else:\n",
    "                    idx_pred=[[t,i-diff[i],j-diff[i]]]\n",
    "                #print 'neg?',j, diff[i],j-diff[i]\n",
    "                #assert j-diff[i]<0\n",
    "            #print j,j-diff[i],diff[i]\n",
    "            elif int(0)>j-diff[i]:\n",
    "                idx_pred=[[t-1,i-diff[i],jn] for jn in range(j)]\n",
    "            else:\n",
    "                idx_pred=[[t-1,i-diff[i],jn] for jn in range(j)]+[[t,i-diff[i],j-diff[i]]]\n",
    "            #print score,coord,idx_pred\n",
    "            #print 'neg?',idx_pred,j-diff[i]\n",
    "            #print score_tpred,score_opred,score\n",
    "            #print [[x,v,w] for x,v,w in idx_pred if score==dfdic[x].loc[v,w]]\n",
    "            predec=[[x,v,w] for x,v,w in idx_pred if score-Spectrum[j]==dfdic[x].loc[v,w]][0]\n",
    "            #print predec\n",
    "            try:\n",
    "                backtrack[str(coord)]=backtrack[str(predec)]+[coord]\n",
    "            except KeyError:\n",
    "                backtrack[str(coord)]=[str(predec)]+[coord]\n",
    "            #print t,i,j\n",
    "            '''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
