{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 1 3 4 6 5 7 8 9 10 11 12 14 13 16 15 18 17 20 19 22 21 24 23 25 26 27 28 29 30 32 31 33 34 36 35 37 38 39 40 42 41 44 43 45 46 47 48 50 49 51 52 53 54 55 56 58 57 60 59 62 61 64 63 66 65 67 68 70 69 71 72 73 74 75 76 77 78 79 80 81 82 84 83 86 85 87 88 90 89 91 92 93 94 96 95 98 97 100 99 102 101 103 104 106 105 108 107 110 109 112 111 114 113 115 116 118 117 120 119 122 121 123 124 126 125 127 128 130 129 132 131 134 133\n"
     ]
    }
   ],
   "source": [
    "''' Implement ChromosomeToCycle. '''\n",
    "'''\n",
    "ChromosomeToCycle(Chromosome)\n",
    "     for j ← 1 to |Chromosome|\n",
    "          i ← Chromosomej\n",
    "          if i > 0\n",
    "               Nodes2j−1 ←2i−1\n",
    "               Nodes2j ← 2i\n",
    "          else\n",
    "               Nodes2j−1 ← -2i\n",
    "               Nodes2j ←-2i−1\n",
    "     return Nodes\n",
    "'''\n",
    "\n",
    "def ChromosomeToCycle(Chromosome):\n",
    "    Nodes={}\n",
    "    for j in range(len(Chromosome)):\n",
    "        i=Chromosome[j]\n",
    "        if i>0:\n",
    "            Nodes[2*j-1] = 2*i-1\n",
    "            Nodes[2*j] = 2*i\n",
    "        else:\n",
    "            Nodes[2*j-1] = -2*i\n",
    "            Nodes[2*j] = -2*i-1 \n",
    "    return [Nodes[key] for key in sorted(Nodes.keys())]\n",
    "\n",
    "### test 1 ###\n",
    "\n",
    "Chromosome='(+1 -2 -3 +4)'\n",
    "\n",
    "#print Chromosome\n",
    "\n",
    "### test 2 ###\n",
    "\n",
    "#Chromosome='(+1 +2 -3 -4 +5 -6 +7 +8 -9 +10 -11 -12 +13 -14 +15 -16 +17 +18 +19 -20 +21 -22 -23 -24 +25 +26 +27 -28 +29 -30 -31 +32 +33 +34 -35 +36 +37 +38 -39 +40 -41 -42 +43 -44 -45 -46 -47 +48 +49 -50 +51 -52 -53 -54 +55 +56 +57 -58 -59 -60 +61 +62 +63 -64 +65 +66 +67 -68 -69)'\n",
    "\n",
    "#####\n",
    "\n",
    "### challenge ###\n",
    "\n",
    "#f=open('../../Downloads/dataset_8222_4.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba6f.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "Chromosome=str(lines[0])\n",
    "\n",
    "#### \n",
    "\n",
    "Chromosome=Chromosome[1:-1]\n",
    "Chromosome=map(int,Chromosome.split(' '))\n",
    "#print Chromosome\n",
    "#print ChromosomeToCycle(Chromosome)\n",
    "print ' '.join(map(str,ChromosomeToCycle(Chromosome)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 1, 2, 4, 3, 6, 5, 8, 7, 9, 10, 12, 11, 14, 13, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 27, 30, 29, 32, 31, 34, 33, 36, 35, 38, 37, 39, 40, 42, 41, 43, 44, 46, 45, 47, 48, 50, 49, 51, 52, 54, 53, 56, 55, 58, 57, 60, 59, 62, 61, 64, 63, 65, 66, 67, 68, 70, 69, 72, 71, 73, 74, 75, 76, 77, 78, 80, 79, 82, 81, 84, 83, 85, 86, 88, 87, 90, 89, 91, 92, 93, 94, 95, 96, 97, 98, 100, 99, 102, 101, 103, 104, 106, 105, 107, 108, 109, 110, 111, 112, 114, 113, 116, 115, 118, 117, 120, 119, 122, 121, 123, 124, 125, 126, 128, 127, 130, 129, 132, 131, 133, 134]\n",
      "(+1 -2 -3 -4 +5 -6 -7 +8 +9 +10 +11 +12 +13 -14 -15 -16 -17 -18 -19 +20 -21 +22 -23 +24 -25 +26 -27 -28 -29 -30 -31 -32 +33 +34 -35 -36 +37 +38 +39 -40 -41 -42 +43 -44 -45 +46 +47 +48 +49 -50 -51 +52 -53 +54 +55 +56 -57 -58 -59 -60 -61 +62 +63 -64 -65 -66 +67)\n"
     ]
    }
   ],
   "source": [
    "'''  Implement CycleToChromosome.'''\n",
    "'''\n",
    "CycleToChromosome(Nodes)\n",
    "     for j ← 1 to |Nodes|/2\n",
    "          if Nodes2j−1 < Nodes2j\n",
    "               Chromosomej ← Nodes2j /2\n",
    "          else\n",
    "               Chromosomej ← −Nodes2j−1/2\n",
    "     return Chromosome\n",
    "'''\n",
    "\n",
    "\n",
    "def printplus(n):\n",
    "    return \"%+d\" % n\n",
    "\n",
    "def CycleToChromosome(Nodes):\n",
    "    Chromosome={}\n",
    "    for j in range (1,((len(Nodes)-1)/2)+1):\n",
    "        if Nodes[2*j-1] < Nodes[2*j]:\n",
    "            Chromosome[j] = \"%+d\" % (Nodes[2*j]/2)\n",
    "        else:\n",
    "            Chromosome[j] = - Nodes[2*j-1]/2\n",
    "    return [Chromosome[key] for key in sorted(Chromosome.keys())]\n",
    "\n",
    "def CycleToChromosome(Nodes):\n",
    "    Chromosome={}\n",
    "    print Nodes\n",
    "    for j in range (1,(len(Nodes)/2)+1):\n",
    "        if Nodes[2*j-1] < Nodes[2*j]:\n",
    "            Chromosome[j] = \"%+d\" % (Nodes[2*j]/2)\n",
    "        else:\n",
    "            Chromosome[j] = - Nodes[2*j-1]/2\n",
    "    return [Chromosome[key] for key in sorted(Chromosome.keys())]\n",
    "\n",
    "\n",
    "### test 1 ###\n",
    "\n",
    "Nodes='(1 2 4 3 6 5 7 8)'\n",
    "\n",
    "### test 2 ###\n",
    "\n",
    "#Nodes='(1 2 3 4 5 6 8 7 9 10 11 12 14 13 15 16 17 18 19 20 21 22 23 24 26 25 27 28 29 30 31 32 33 34 36 35 37 38 39 40 41 42 43 44 45 46 48 47 49 50 52 51 54 53 56 55 57 58 60 59 61 62 63 64 65 66 68 67 70 69 72 71 73 74 76 75 77 78 79 80 81 82 83 84 85 86 87 88 89 90 92 91 93 94 96 95 97 98 100 99 102 101 103 104 106 105 108 107 109 110 111 112 113 114 115 116 118 117 120 119 122 121 123 124 125 126 128 127 130 129 132 131 133 134 136 135 137 138 140 139)'\n",
    "\n",
    "### challenge ###\n",
    "\n",
    "#f=open('../../Downloads/dataset_8222_5.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba6g.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "Nodes=str(lines[0])\n",
    "\n",
    "###\n",
    "\n",
    "Nodes='0 '+Nodes[1:-1]\n",
    "#print Nodes\n",
    "Nodes=map(int,Nodes.split(' '))\n",
    "#print Nodes \n",
    "#print CycleToChromosome(Nodes)\n",
    "print '('+' '.join(map(str,CycleToChromosome(Nodes)))+')'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "((2, 3), (4, 5), (6, 8), (7, 10), (9, 12), (11, 13), (14, 16), (15, 18), (17, 19), (20, 21), (22, 23), (24, 26), (25, 28), (27, 30), (29, 31), (32, 34), (33, 36), (35, 37), (38, 39), (40, 42), (41, 43), (44, 46), (45, 48), (47, 50), (49, 51), (52, 54), (53, 55), (56, 58), (57, 1), (59, 62), (61, 64), (63, 66), (65, 67), (68, 69), (70, 72), (71, 73), (74, 76), (75, 77), (78, 80), (79, 81), (82, 84), (83, 85), (86, 88), (87, 90), (89, 92), (91, 94), (93, 95), (96, 98), (97, 100), (99, 101), (102, 104), (103, 105), (106, 108), (107, 110), (109, 111), (112, 113), (114, 60), (115, 118), (117, 119), (120, 122), (121, 124), (123, 126), (125, 128), (127, 130), (129, 132), (131, 134), (133, 135), (136, 138), (137, 139), (140, 141), (142, 144), (143, 146), (145, 147), (148, 150), (149, 151), (152, 154), (153, 156), (155, 157), (158, 159), (160, 161), (162, 163), (164, 165), (166, 167), (168, 170), (169, 172), (171, 173), (174, 116), (176, 178), (177, 179), (180, 181), (182, 184), (183, 186), (185, 188), (187, 189), (190, 191), (192, 193), (194, 196), (195, 198), (197, 200), (199, 202), (201, 203), (204, 206), (205, 207), (208, 209), (210, 211), (212, 214), (213, 215), (216, 217), (218, 220), (219, 222), (221, 175), (223, 225), (226, 228), (227, 229), (230, 232), (231, 233), (234, 235), (236, 238), (237, 240), (239, 241), (242, 243), (244, 246), (245, 248), (247, 250), (249, 251), (252, 253), (254, 255), (256, 257), (258, 259), (260, 261), (262, 263), (264, 266), (265, 224), (268, 270), (269, 272), (271, 274), (273, 275), (276, 277), (278, 279), (280, 282), (281, 283), (284, 286), (285, 287), (288, 289), (290, 291), (292, 293), (294, 295), (296, 298), (297, 299), (300, 301), (302, 303), (304, 306), (305, 308), (307, 267), (310, 311), (312, 314), (313, 316), (315, 318), (317, 319), (320, 322), (321, 324), (323, 325), (326, 327), (328, 330), (329, 331), (332, 334), (333, 336), (335, 337), (338, 339), (340, 342), (341, 344), (343, 345), (346, 347), (348, 349), (350, 352), (351, 353), (354, 356), (355, 309), (358, 359), (360, 361), (362, 364), (363, 365), (366, 367), (368, 369), (370, 371), (372, 373), (374, 376), (375, 377), (378, 379), (380, 382), (381, 384), (383, 385), (386, 388), (387, 389), (390, 391), (392, 393), (394, 396), (395, 397), (398, 400), (399, 402), (401, 404), (403, 406), (405, 408), (407, 409), (410, 411), (412, 413), (414, 415), (416, 357))\n"
     ]
    }
   ],
   "source": [
    "''' Implement ColoredEdges. '''\n",
    "''' \n",
    "ColoredEdges(P)\n",
    "     Edges ← an empty set\n",
    "     for each chromosome Chromosome in P\n",
    "          Nodes ← ChromosomeToCycle(Chromosome)\n",
    "          for j ← 1 to |Chromosome|\n",
    "               add the edge (Nodes2j, Nodes2j +1) to Edges\n",
    "     return Edges\n",
    "'''\n",
    "\n",
    "def ColoredEdges(P):\n",
    "    Edges =[]\n",
    "    for Chromosome in P:\n",
    "        Chromosome=map(int,Chromosome.split(' '))\n",
    "        Nodes = [0]+ChromosomeToCycle(Chromosome)+[ChromosomeToCycle(Chromosome)[0]]\n",
    "        for j in range (1 , len(Chromosome)+1):\n",
    "            Edges.append([Nodes[2*j],Nodes[2*j+1]])\n",
    "    return Edges\n",
    "\n",
    "### test 1 ###\n",
    "\n",
    "P='(+1 -2 -3)(+4 +5 -6)'\n",
    "\n",
    "### test 2 ###\n",
    "\n",
    "#P='(+1 -2 -3 +4 -5 -6 -7 -8 +9 -10 -11 -12 +13 +14 +15 -16 -17 +18 +19 +20 +21)(-22 -23 -24 -25 -26 -27 -28 +29 +30 +31 -32 -33 +34 +35 -36 +37 +38 +39 -40 +41 +42 +43 +44)(+45 +46 +47 +48 -49 +50 -51 +52 -53 +54 +55 -56 +57 -58 +59 -60 -61 -62 -63 -64 +65 -66)(-67 +68 +69 +70 -71 -72 +73 +74 -75 +76 -77 -78 +79 +80 +81 +82 -83 +84 +85 +86 -87 -88 -89 +90)(-91 -92 +93 +94 -95 -96 +97 +98 -99 -100 -101 -102 +103 +104 -105 +106 -107 -108 +109 +110 +111 -112 +113 +114 -115 +116 +117 -118 +119)(-120 +121 +122 -123 -124 -125 +126 +127 +128 +129 -130 +131 -132 +133 -134 -135 -136 -137 +138 -139 +140 +141 +142 +143 -144 -145 +146 -147 +148)(-149 -150 +151 -152 -153 -154 -155 -156 +157 +158 +159 +160 -161 +162 +163 +164 +165 -166 +167 -168 -169 +170 -171 +172)'\n",
    "\n",
    "### challenge ###\n",
    "\n",
    "#f=open('../../Downloads/dataset_8222_7.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba6h.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "P=str(lines[0])\n",
    "\n",
    "#####\n",
    "\n",
    "P=P[1:-1]\n",
    "P=P.split(')(')\n",
    "#print P\n",
    "#print ColoredEdges(P)\n",
    "print str(ColoredEdges(P)).replace('[','(').replace(']',')')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(+1 +29 -28 -27 -26 -25 +24 +23 -22 +21 +20 -19 +18 -17 +16 +15 -14 -13 -12 -11 -10 +9 +8 +7 -6 +5 +4 +3 +2)(-30 +31 +32 +33 -34 +35 +36 -37 -38 +39 -40 +41 +42 -43 +44 +45 -46 +47 -48 -49 +50 -51 -52 -53 +54)(+55 -56 +57 +58 -59 -60 +61 -62 -63 -64 +65 -66 +67 +68 -69 -70 -71 +72 -73 +74 +75 +76 -77 +78 -79 -80 -81 +82)(+83 +84 +85 +86 -87 +88 +89 +90 +91 +92 +93 +94 -95 +96 +97 -98 -99 +100 -101 +102 +103 -104)(+105 +106 -107 +108 +109 +110 -111 +112 +113 +114 +115 +116 -117 +118 -119 -120 -121 +122 +123 -124 -125 -126 +127)(-128 +129 +130 +131 -132 +133 -134 +135 +136 +137 -138 +139 -140 +141 +142 +143 -144 +145 -146 -147 +148 +149 -150 -151 +152 +153 +154 -155 -156 +157)(-158 -159 -160 +161 -162 -163 +164 +165 +166 +167 +168 +169 -170 +171 +172 +173 +174 -175 +176 -177 -178 -179 +180 -181 -182 +183 +184 +185)(+186 -187 -188 +189 -190 +191 -192 -193 +194 +195 -196 +197 +198 -199 -200 +201 +202 +203 -204 +205 +206 -207 -208 +209 -210 -211)(-212 -213 -214 -215 +216 -217 +218 -219 +220 +221 -222 +223 -224 +225 +226 +227 -228 +229 -230 -231 -232 +233 -234 -235 -236 +237 +238)\n"
     ]
    }
   ],
   "source": [
    "'''Implement GraphToGenome.'''\n",
    "'''\n",
    "GraphToGenome(GenomeGraph)\n",
    "     P ← an empty set of chromosomes\n",
    "     for each cycle Nodes in GenomeGraph\n",
    "          Nodes﻿ ← sequence of nodes in this cycle (starting from node 1)\n",
    "          Chromosome ← CycleToChromosome(Nodes)\n",
    "          add Chromosome to P\n",
    "     return P\n",
    "'''\n",
    "\n",
    "#no good\n",
    "'''def GraphToGenome(ColoredEdges):\n",
    "    Genome=[]\n",
    "    GenomeCycles=[]\n",
    "    ChrCycle=[]\n",
    "    for edge in ColoredEdges:\n",
    "        ChrCycle+=edge\n",
    "        if edge[0]>edge[1]:\n",
    "            ChrCycle = [ChrCycle.pop(-1)]+ChrCycle\n",
    "            GenomeCycles+=[ChrCycle]\n",
    "            ChrCycle=[]\n",
    "\n",
    "    for cycle in GenomeCycles:\n",
    "        cycle=[0]+cycle\n",
    "        Genome+=[CycleToChromosome(cycle)]\n",
    "    return Genome'''\n",
    "##########################3\n",
    "\n",
    "def graph_to_genome(g):\n",
    "    '''\n",
    "    CODE CHALLENGE: Implement GraphToGenome.\n",
    "    Input: The colored edges ColoredEdges of a genome graph.\n",
    "    Output: The genome P corresponding to this genome graph.\n",
    "    '''\n",
    "    import numpy as np\n",
    "    genome = []\n",
    "    visit = []\n",
    "    adj = np.zeros(len(g)*2, dtype = np.int)\n",
    "    for t in g:\n",
    "        adj[t[0]-1] = t[1]-1\n",
    "        adj[t[1]-1] = t[0]-1\n",
    "    \n",
    "    for t in g:\n",
    "        orig = t[0]\n",
    "        if orig in visit:\n",
    "            continue\n",
    "        visit.append(orig)\n",
    "        if (orig%2 == 0):\n",
    "            closing = orig-1\n",
    "        else:\n",
    "            closing = orig+1\n",
    "        p = []\n",
    "        i = 0\n",
    "        while(True):\n",
    "            if (orig%2 == 0):\n",
    "                p.append(orig/2)\n",
    "            else:\n",
    "                p.append(-(orig+1)/2)\n",
    "            dest = adj[orig-1]+1\n",
    "            i = i + 1\n",
    "            if (i>100):\n",
    "                assert False\n",
    "                return\n",
    "            visit.append(dest)\n",
    "            if (dest == closing):\n",
    "                genome.append(p)\n",
    "                break\n",
    "            if (dest%2 == 0):\n",
    "                orig = dest -1\n",
    "            else:\n",
    "                orig = dest + 1\n",
    "            assert orig > 0\n",
    "            visit.append(orig)\n",
    "    return genome\n",
    "###################################\n",
    "\n",
    "### test 1 ###\n",
    "\n",
    "ColoredEdges=(2, 4), (3, 6), (5, 1), (7, 9), (10, 12), (11, 8)\n",
    "ColoredEdges=[list(item) for item in list(ColoredEdges)]\n",
    "\n",
    "### test 2 ###\n",
    "\n",
    "#ColoredEdges=(1, 4), (3, 6), (5, 7), (8, 9), (10, 11), (12, 14), (13, 15), (16, 18), (17, 19), (20, 22), (21, 24), (23, 26), (25, 27), (28, 29), (30, 32), (31, 33), (34, 36), (35, 38), (37, 39), (40, 42), (41, 44), (43, 2), (45, 48), (47, 50), (49, 51), (52, 54), (53, 56), (55, 57), (58, 60), (59, 62), (61, 63), (64, 66), (65, 68), (67, 69), (70, 72), (71, 74), (73, 76), (75, 78), (77, 79), (80, 81), (82, 84), (83, 86), (85, 87), (88, 90), (89, 91), (92, 94), (93, 46), (96, 98), (97, 99), (100, 102), (101, 103), (104, 106), (105, 107), (108, 109), (110, 111), (112, 114), (113, 115), (116, 117), (118, 119), (120, 122), (121, 123), (124, 125), (126, 127), (128, 130), (129, 131), (132, 134), (133, 135), (136, 138), (137, 139), (140, 141), (142, 144), (143, 145), (146, 147), (148, 149), (150, 152), (151, 95), (154, 155), (156, 158), (157, 159), (160, 161), (162, 163), (164, 166), (165, 168), (167, 169), (170, 171), (172, 174), (173, 175), (176, 178), (177, 180), (179, 182), (181, 184), (183, 186), (185, 187), (188, 190), (189, 191), (192, 194), (193, 196), (195, 153), (198, 200), (199, 201), (202, 204), (203, 205), (206, 207), (208, 210), (209, 211), (212, 214), (213, 215), (216, 217), (218, 219), (220, 222), (221, 223), (224, 226), (225, 227), (228, 230), (229, 232), (231, 234), (233, 235), (236, 238), (237, 239), (240, 241), (242, 244), (243, 245), (246, 197), (248, 250), (249, 252), (251, 253), (254, 255), (256, 258), (257, 260), (259, 262), (261, 263), (264, 265), (266, 267), (268, 270), (269, 272), (271, 273), (274, 276), (275, 278), (277, 280), (279, 281), (282, 283), (284, 286), (285, 247), (288, 290), (289, 292), (291, 293), (294, 295), (296, 297), (298, 299), (300, 302), (301, 303), (304, 306), (305, 307), (308, 310), (309, 312), (311, 314), (313, 316), (315, 318), (317, 319), (320, 322), (321, 323), (324, 325), (326, 327), (328, 330), (329, 332), (331, 333), (334, 336), (335, 338), (337, 339), (340, 287)\n",
    "#ColoredEdges=[list(item) for item in list(ColoredEdges)]\n",
    "\n",
    "### challenge ###\n",
    "\n",
    "#f=open('../../Downloads/dataset_8222_8.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba6i(1).txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "pairs=[line.replace('(','').replace(' ','').split(',') for line in lines[0][0:-1].split('),')]\n",
    "\n",
    "for pair in pairs:\n",
    "    pair=map(int,pair)\n",
    "    ColoredEdges+=[pair]\n",
    "\n",
    "#ColoredEdges=(1, 4), (3, 6), (5, 8), (7, 10), (9, 12), (11, 13), (14, 16), (15, 17), (18, 20), (19, 22), (21, 24), (23, 26), (25, 27), (28, 29), (30, 31), (32, 34), (33, 36), (35, 37), (38, 40), (39, 42), (41, 43), (44, 45), (46, 2), (48, 49), (50, 51), (52, 54), (53, 56), (55, 57), (58, 59), (60, 61), (62, 63), (64, 65), (66, 67), (68, 70), (69, 71), (72, 74), (73, 75), (76, 78), (77, 79), (80, 81), (82, 84), (83, 85), (86, 87), (88, 90), (89, 91), (92, 47), (94, 96), (95, 97), (98, 100), (99, 102), (101, 104), (103, 106), (105, 108), (107, 110), (109, 112), (111, 114), (113, 115), (116, 117), (118, 120), (119, 121), (122, 124), (123, 126), (125, 127), (128, 129), (130, 131), (132, 134), (133, 93), (136, 137), (138, 139), (140, 141), (142, 143), (144, 146), (145, 148), (147, 150), (149, 152), (151, 154), (153, 155), (156, 158), (157, 160), (159, 161), (162, 164), (163, 166), (165, 168), (167, 169), (170, 171), (172, 173), (174, 175), (176, 177), (178, 180), (179, 182), (181, 184), (183, 185), (186, 188), (187, 135), (189, 192), (191, 194), (193, 195), (196, 198), (197, 200), (199, 202), (201, 204), (203, 206), (205, 207), (208, 210), (209, 212), (211, 213), (214, 215), (216, 218), (217, 219), (220, 221), (222, 224), (223, 226), (225, 227), (228, 229), (230, 232), (231, 234), (233, 236), (235, 237), (238, 240), (239, 241), (242, 244), (243, 246), (245, 248), (247, 190), (249, 252), (251, 253), (254, 255), (256, 258), (257, 260), (259, 261), (262, 264), (263, 265), (266, 267), (268, 270), (269, 271), (272, 274), (273, 275), (276, 277), (278, 279), (280, 281), (282, 283), (284, 285), (286, 288), (287, 290), (289, 291), (292, 294), (293, 295), (296, 297), (298, 250), (300, 302), (301, 303), (304, 305), (306, 307), (308, 309), (310, 311), (312, 314), (313, 316), (315, 317), (318, 320), (319, 322), (321, 323), (324, 326), (325, 327), (328, 330), (329, 331), (332, 333), (334, 335), (336, 337), (338, 339), (340, 341), (342, 344), (343, 299)\n",
    "#print ColoredEdges\n",
    "\n",
    "###\n",
    "#print GraphToGenome(ColoredEdges)\n",
    "#print graph_to_genome(ColoredEdges)\n",
    "#print '('+')('.join([' '.join(map(str,item)) for item in GraphToGenome(ColoredEdges)])+')'\n",
    "\n",
    "output=[]\n",
    "for chrom in graph_to_genome(ColoredEdges):\n",
    "    output.append('('+' '.join([\"%+d\" % node for node in chrom])+')')\n",
    "print ''.join(output)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 3), (4, 5), (6, 8), (7, 10), (9, 11), (12, 14), (13, 15), (16, 18), (17, 19), (20, 22), (21, 23), (24, 25), (26, 28), (27, 29), (30, 32), (31, 33), (34, 35), (36, 38), (37, 40), (39, 41), (42, 43), (44, 46), (45, 47), (48, 49), (50, 52), (51, 54), (53, 56), (55, 57), (58, 60), (59, 61), (62, 64), (63, 66), (65, 67), (68, 70), (69, 71), (72, 73), (74, 75), (76, 78), (77, 79), (80, 82), (81, 83), (84, 86), (85, 88), (87, 90), (89, 91), (92, 94), (93, 96), (95, 97), (98, 99), (100, 102), (101, 103), (104, 106), (105, 108), (107, 110), (109, 111), (112, 113), (114, 116), (115, 117), (118, 120), (119, 122), (121, 123), (124, 126), (125, 127), (128, 129), (130, 131), (132, 134), (133, 136), (135, 138), (137, 2), (36, 34), (38, 35)\n"
     ]
    }
   ],
   "source": [
    "'''Implement 2-BreakOnGenomeGraph'''\n",
    "'''\n",
    "2-BreakOnGenomeGraph(GenomeGraph, i, i′, j, j′)\n",
    "     remove colored edges (i, i') and (j, j′) from GenomeGraph\n",
    "     add colored edges (i, j) and (i′, j') to GenomeGraph\n",
    "     return GenomeGraph\n",
    "'''\n",
    " # no good\n",
    "    \n",
    "'''def TwoBreakOnGenomeGraph(GenomeGraph, i, ii, j, jj):\n",
    "    if (i,ii) in GenomeGraph:\n",
    "        GenomeGraph.remove((i,ii))\n",
    "    else:\n",
    "        GenomeGraph.remove((ii,i))\n",
    "    if (j,jj) in GenomeGraph:\n",
    "        GenomeGraph.remove((j,jj))\n",
    "    else:\n",
    "        GenomeGraph.remove((jj,j))\n",
    "    \n",
    "    GenomeGraph.append((i,j))\n",
    "    GenomeGraph.append((ii,jj))\n",
    "    return GenomeGraph'''\n",
    "\n",
    "\n",
    "\n",
    "## worked\n",
    "def two_break_on_genome_graph(g,i,j,k,l):\n",
    "    bg = []\n",
    "    rem = ((i,j), (j,i), (k,l), (l,k))\n",
    "    bg = [ t for t in g if t not in rem]\n",
    "    bg.append((i,k))\n",
    "    bg.append((j,l))\n",
    "    \n",
    "    return bg\n",
    "\n",
    "### test 1 ###\n",
    "\n",
    "GenomeGraph=(2, 4), (3, 8), (7, 5), (6, 1)\n",
    "i,ii,j,jj=1, 6, 3, 8\n",
    "\n",
    "### test 2 ###\n",
    "\n",
    "GenomeGraph=(2, 4), (3, 5), (6, 8), (7, 10), (9, 12), (11, 13), (14, 15), (16, 17), (18, 19), (20, 21), (22, 24), (23, 25), (26, 28), (27, 30), (29, 32), (31, 33), (34, 35), (36, 37), (38, 40), (39, 42), (41, 44), (43, 45), (46, 47), (48, 49), (50, 51), (52, 53), (54, 56), (55, 58), (57, 59), (60, 61), (62, 64), (63, 66), (65, 68), (67, 69), (70, 72), (71, 73), (74, 75), (76, 77), (78, 80), (79, 81), (82, 83), (84, 86), (85, 88), (87, 90), (89, 92), (91, 94), (93, 96), (95, 98), (97, 99), (100, 102), (101, 104), (103, 106), (105, 107), (108, 109), (110, 111), (112, 113), (114, 116), (115, 117), (118, 120), (119, 121), (122, 124), (123, 126), (125, 128), (127, 129), (130, 132), (131, 134), (133, 136), (135, 1)\n",
    "i,ii,j,jj=87, 90, 74, 75\n",
    "\n",
    "### challenge ###\n",
    "\n",
    "#f=open('../../Downloads/dataset_8224_2.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba6j(1).txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "i,ii,j,jj=map(int,lines[1].split(', '))\n",
    "pairs=[line.replace('(','').replace(' ','').split(',') for line in lines[0][0:-1].split('),')]\n",
    "GenomeGraph=[]\n",
    "for pair in pairs:\n",
    "    pair=map(int,pair)\n",
    "    GenomeGraph+=[pair]\n",
    "\n",
    "#print GenomeGraph, i,ii,j,jj\n",
    "###\n",
    "\n",
    "\n",
    "#print TwoBreakOnGenomeGraph(GenomeGraph, i, ii, j, jj)\n",
    "#print two_break_on_genome_graph(GenomeGraph, i, ii, j, jj)\n",
    "result=str(two_break_on_genome_graph(GenomeGraph, i, ii, j, jj))\n",
    "print result.replace('[','(').replace(']',')')[1:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2, 4), (3, 6), (5, 8), (7, 9), (10, 12), (11, 14), (13, 15), (16, 17), (18, 19), (20, 22), (21, 23), (24, 26), (25, 28), (27, 29), (30, 32), (31, 33), (34, 36), (35, 38), (37, 40), (39, 42), (41, 43), (44, 45), (46, 47), (48, 50), (49, 51), (52, 53), (54, 55), (56, 57), (58, 60), (59, 62), (61, 63), (64, 66), (65, 68), (67, 69), (70, 71), (72, 73), (74, 75), (76, 78), (77, 80), (79, 82), (81, 84), (83, 86), (85, 88), (87, 90), (89, 92), (91, 94), (93, 96), (95, 97), (98, 100), (99, 102), (101, 103), (104, 106), (105, 107), (108, 109), (110, 111), (112, 114), (113, 115), (116, 117), (118, 120), (119, 122), (121, 1)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: ')'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-170-96749e1f9632>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    134\u001b[0m \u001b[1;32mprint\u001b[0m \u001b[0mP\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    135\u001b[0m \u001b[1;31m##\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 136\u001b[1;33m \u001b[0mP\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mmap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mint\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mitem\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m','\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m''\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mitem\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mP\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m' '\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    137\u001b[0m \u001b[1;31m#print P\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    138\u001b[0m \u001b[1;31m#print n,nn,o,oo\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: invalid literal for int() with base 10: ')'"
     ]
    }
   ],
   "source": [
    "#### NO GOOD ####\n",
    "\n",
    "'''Implement 2-BreakOnGenome.'''\n",
    "'''\n",
    "2-BreakOnGenome(P, i, i′, j, j′)\n",
    "     GenomeGraph ← BlackEdges(P) and ColoredEdges(P)\n",
    "     GenomeGraph ← 2-BreakOnGenome(GenomeGraph, i, i′, j, j′)\n",
    "     P ← GraphToGenome(GenomeGraph)\n",
    "     return P\n",
    "'''\n",
    "def ChromosomeToCycle(Chromosome):\n",
    "    Nodes={}\n",
    "    for j in range(len(Chromosome)):\n",
    "        i=Chromosome[j]\n",
    "        if i>0:\n",
    "            Nodes[2*j-1] = 2*i-1\n",
    "            Nodes[2*j] = 2*i\n",
    "        else:\n",
    "            Nodes[2*j-1] = -2*i\n",
    "            Nodes[2*j] = -2*i-1 \n",
    "    return [Nodes[key] for key in sorted(Nodes.keys())]\n",
    "\n",
    "'''\n",
    "def CycleToChromosome(Nodes):\n",
    "    Chromosome={}\n",
    "    for j in range (1,((len(Nodes)-1)/2)+1):\n",
    "        if Nodes[2*j-1] < Nodes[2*j]:\n",
    "            Chromosome[j] = \"%+d\" % (Nodes[2*j]/2)\n",
    "        else:\n",
    "            Chromosome[j] = - Nodes[2*j-1]/2\n",
    "    return [Chromosome[key] for key in sorted(Chromosome.keys())]\n",
    "'''\n",
    "\n",
    "def CycleToChromosome(Nodes):\n",
    "    Chromosome={}\n",
    "    for j in range (1,(len(Nodes)/2)):\n",
    "        if Nodes[2*j-1] < Nodes[2*j]:\n",
    "            Chromosome[j] = \"%+d\" % (Nodes[2*j]/2)\n",
    "        else:\n",
    "            Chromosome[j] = - Nodes[2*j-1]/2\n",
    "    return [Chromosome[key] for key in sorted(Chromosome.keys())]\n",
    "\n",
    "\n",
    "def GraphToGenome(ColoredEdges):\n",
    "    \n",
    "    breakpointcycles=[]\n",
    "    while len(ColoredEdges) >0:\n",
    "        i=0\n",
    "        cycle=[]\n",
    "        cycle+=ColoredEdges.pop(i)\n",
    "        while cycle[0] != cycle[-1]:\n",
    "            for j in range(len(ColoredEdges)):\n",
    "                edge=ColoredEdges[j]\n",
    "                if cycle[-1] in edge:\n",
    "                    edge=ColoredEdges.pop(j)\n",
    "                    \n",
    "                    edge.remove(cycle[-1])\n",
    "                    cycle=cycle+edge\n",
    "                    break\n",
    "        cycle=cycle[:-1]\n",
    "        while cycle[-1]*2< cycle[-2]:\n",
    "            cycle=[cycle.pop(-1)]+cycle\n",
    "        #print cycle\n",
    "        breakpointcycles+=[[0]+cycle+[cycle[1]]]\n",
    "    return [CycleToChromosome(Chr) for Chr in breakpointcycles]\n",
    "\n",
    "\n",
    "def BlackEdges(Chromosome):\n",
    "    Nodes={}\n",
    "    for j in range(len(Chromosome)):\n",
    "        i=Chromosome[j]\n",
    "        if i>0:\n",
    "            Nodes[2*j-1] = 2*i-1\n",
    "            Nodes[2*j] = 2*i\n",
    "        else:\n",
    "            Nodes[2*j-1] = -2*i\n",
    "            Nodes[2*j] = -2*i-1 \n",
    "    theList=[Nodes[key] for key in sorted(Nodes.keys())]\n",
    "    N=2\n",
    "    return [theList[n:n+N] for n in range(0, len(theList), N)]\n",
    "\n",
    "def ColoredEdges(Chromosome):\n",
    "    Edges =[]\n",
    "        #Chromosome=map(int,Chromosome.split(' '))\n",
    "    Nodes = [0]+ChromosomeToCycle(Chromosome)+[ChromosomeToCycle(Chromosome)[0]]\n",
    "    for j in range (1 , len(Chromosome)+1):\n",
    "        Edges.append([Nodes[2*j],Nodes[2*j+1]])\n",
    "    return Edges\n",
    "\n",
    "def TwoBreakOnGenomeGraph(GenomeGraph, i, ii, j, jj):\n",
    "    #print [i,ii]\n",
    "    if [i,ii] in GenomeGraph:\n",
    "        GenomeGraph.remove([i,ii])\n",
    "    else:\n",
    "        GenomeGraph.remove([ii,i])\n",
    "    if [j,jj] in GenomeGraph:\n",
    "        GenomeGraph.remove([j,jj])\n",
    "    else:\n",
    "        GenomeGraph.remove([jj,j])\n",
    "    \n",
    "    GenomeGraph.append([i,j])\n",
    "    GenomeGraph.append([ii,jj])\n",
    "    return GenomeGraph\n",
    "\n",
    "def TwoBreakOnGenome(P, i, ii, j, jj):\n",
    "    GenomeGraph = ColoredEdges(P) + BlackEdges(P)\n",
    "    #print GenomeGraph\n",
    "    GenomeGraph = sorted(TwoBreakOnGenomeGraph(GenomeGraph, i, ii, j, jj))\n",
    "    #print GenomeGraph\n",
    "    P=GraphToGenome(GenomeGraph)\n",
    "    return P\n",
    "\n",
    "\n",
    "### test 1 ###\n",
    "\n",
    "P='(+1 -2 -4 +3)'\n",
    "#P=map(int,P[1:-1].split(' '))\n",
    "n,nn,o,oo= 1, 6, 3, 8\n",
    "\n",
    "### test 2 ###\n",
    "'''\n",
    "P='(-1 +2 -3 -4 -5 -6 -7 +8 +9 +10 +11 +12 +13 +14 +15 -16 -17 -18 +19 -20 +21 +22 -23 -24 +25 -26 -27 +28 -29 -30 +31 +32 +33 -34 -35 +36 +37 +38 +39 -40 +41 +42 -43 +44 +45 +46 -47 +48 +49 +50 -51 +52 +53 +54 -55 -56 -57 +58 -59 -60 -61 +62 -63 +64 +65 -66 -67)'\n",
    "#P=map(int,P[1:-1].split(' '))\n",
    "n,nn,o,oo= 116, 118, 73, 72 \n",
    "'''\n",
    "\n",
    "### challenge ###\n",
    "\n",
    "#f=open('../../Downloads/dataset_8224_3.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba6j.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "P=lines[0]\n",
    "n,nn,o,oo=map(int,lines[1].split(', '))\n",
    "\n",
    "print P\n",
    "##\n",
    "P=[map(int,item.replace(',','')) for item in P[1:-1].split(' ')]\n",
    "#print P\n",
    "#print n,nn,o,oo\n",
    "#print ColoredEdges(P)\n",
    "#print BlackEdges(P)+ColoredEdges(P)\n",
    "\n",
    "#print TwoBreakOnGenome(P, n, nn, o, oo)\n",
    "print '('+')('.join([' '.join(map(str,item)) for item in TwoBreakOnGenome(P, n, nn, o, oo)]  )+')'\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(+1 -2 -3 -4 -5 +6 -7 -8 -9 -10 -11 -12 +13 +14 -15 +16 +17 -18 -22 +23 +24 +25 -26 -27 -28 +29 +30 +31 -32 +33 +34 +35 -36 -37 +38 +39 +40 -41 +42 -43 -44 -45 +46 +47 +48 -49 +50 -51 +52 -53 +54 -55 +56 -57 +58 +59 -60 +61 -62 -63 -64 -65 -66) (+19 -20 -21)\n"
     ]
    }
   ],
   "source": [
    "'''Implement 2-BreakOnGenome.'''\n",
    "'''\n",
    "2-BreakOnGenome(P, i, i′, j, j′)\n",
    "     GenomeGraph ← BlackEdges(P) and ColoredEdges(P)\n",
    "     GenomeGraph ← 2-BreakOnGenome(GenomeGraph, i, i′, j, j′)\n",
    "     P ← GraphToGenome(GenomeGraph)\n",
    "     return P\n",
    "'''\n",
    "\n",
    "def chromosome_to_cycle(p):\n",
    "    nodes = []\n",
    "    for i in p:\n",
    "        if (i>0):\n",
    "            nodes.append(2*i-1)\n",
    "            nodes.append(2*i)\n",
    "        else:\n",
    "            nodes.append(-2*i)\n",
    "            nodes.append(-2*i-1)\n",
    "    return nodes\n",
    "\n",
    "\n",
    "\n",
    "def cycle_to_chromosome(nodes):\n",
    "    p = []\n",
    "    for j in range(0,len(nodes)/2):\n",
    "        if nodes[2*j] < nodes[2*j+1]:\n",
    "            s = j+1\n",
    "        else:\n",
    "            s = -(j+1)\n",
    "        p.append(s)\n",
    "    return p\n",
    "\n",
    "def colored_edges(genome):\n",
    "    g = []\n",
    "    for p in genome:\n",
    "        s = chromosome_to_cycle(p)\n",
    "        for j in range(len(s)/2):\n",
    "            head = 1+2*j\n",
    "            tail = (2+2*j) % len(s)\n",
    "            e = (s[head],s[tail])\n",
    "            g.append(e)\n",
    "    return g\n",
    "\n",
    "def graph_to_genome(g):\n",
    "    import numpy as np\n",
    "    genome = []\n",
    "    visit = []\n",
    "    adj = np.zeros(len(g)*2, dtype = np.int)\n",
    "    for t in g:\n",
    "        adj[t[0]-1] = t[1]-1\n",
    "        adj[t[1]-1] = t[0]-1\n",
    "    \n",
    "    for t in g:\n",
    "        orig = t[0]\n",
    "        if orig in visit:\n",
    "            continue\n",
    "        visit.append(orig)\n",
    "        if (orig%2 == 0):\n",
    "            closing = orig-1\n",
    "        else:\n",
    "            closing = orig+1\n",
    "        p = []\n",
    "        i = 0\n",
    "        while(True):\n",
    "            if (orig%2 == 0):\n",
    "                p.append(orig/2)\n",
    "            else:\n",
    "                p.append(-(orig+1)/2)\n",
    "            dest = adj[orig-1]+1\n",
    "            i = i + 1\n",
    "            if (i>100):\n",
    "                assert False\n",
    "                return\n",
    "            visit.append(dest)\n",
    "            if (dest == closing):\n",
    "                genome.append(p)\n",
    "                break\n",
    "            if (dest%2 == 0):\n",
    "                orig = dest -1\n",
    "            else:\n",
    "                orig = dest + 1\n",
    "            assert orig > 0\n",
    "            visit.append(orig)\n",
    "    return genome\n",
    "\n",
    "\n",
    "\n",
    "def two_break_dist(P, Q):\n",
    "    graph = defaultdict(list)\n",
    "    for perm_cycle in P+Q:\n",
    "        n = len(perm_cycle)\n",
    "        for i in xrange(n):\n",
    "            graph[perm_cycle[i]].append(-1*perm_cycle[(i+1) % n])\n",
    "            graph[-1*perm_cycle[(i+1) % n]].append(perm_cycle[i])\n",
    "    component_count = 0\n",
    "    remaining = set(graph.keys())\n",
    "    while remaining:\n",
    "        component_count += 1\n",
    "        queue = {remaining.pop()}  # Undirected graph, so we can choose a remaining node arbitrarily.\n",
    "        while queue:\n",
    "            current = queue.pop()\n",
    "            new_nodes = {node for node in graph[current] if node in remaining}\n",
    "            queue |= new_nodes\n",
    "            remaining -= new_nodes\n",
    "    return sum(map(len,P)) - component_count\n",
    "\n",
    "def colored_edges_cycles(blue, red):\n",
    "    import numpy as np\n",
    "    cycles = []\n",
    "    size = len(blue)+len(red) \n",
    "    adj = np.zeros(shape = (size,2), dtype = np.int)\n",
    "    visited = np.zeros(shape = size, dtype = np.bool)\n",
    "    for e in blue:\n",
    "        adj[e[0]-1,0] = e[1]-1\n",
    "        adj[e[1]-1,0] = e[0]-1\n",
    "    for e in red:\n",
    "        adj[e[0]-1,1] = e[1]-1\n",
    "        adj[e[1]-1,1] = e[0]-1\n",
    "    \n",
    "    for node in range(size):\n",
    "        if not visited[node]:\n",
    "            visited[node] = True\n",
    "            head = node\n",
    "            cycle = [head+1]\n",
    "            # arbitrary we start with a blue edge\n",
    "            color = 0\n",
    "            while (True):\n",
    "                node = adj[node,color]\n",
    "                if (node == head):\n",
    "                    # cycle found, we got back to the visited head node, \n",
    "                    # just break\n",
    "                    cycles.append(cycle)\n",
    "                    break\n",
    "                cycle.append(node+1)\n",
    "                visited[node] = True\n",
    "                color = (color+1) % 2\n",
    "    return cycles\n",
    "\n",
    "def two_break_on_genome_graph(g,i,j,k,l):\n",
    "    bg = []\n",
    "    rem = ((i,j), (j,i), (k,l), (l,k))\n",
    "    bg = [ t for t in g if t not in rem]\n",
    "    bg.append((i,k))\n",
    "    bg.append((j,l))\n",
    "    \n",
    "    return bg\n",
    "\n",
    "def two_break_on_genome(genome,i,j,k,l):\n",
    "    '''\n",
    "    CODE CHALLENGE: Implement 2-BreakOnGenome.\n",
    "    Input: A genome P, followed by indices i, i', j, and j'.\n",
    "    Output: The genome P' resulting from applying the 2-break operation\n",
    "    2-BreakOnGenomeGraph(GenomeGraph, i, i′, j, j′).\n",
    "    '''\n",
    "    g = colored_edges(genome)\n",
    "    g = two_break_on_genome_graph(g,i,j,k,l)\n",
    "    genome = graph_to_genome(g)\n",
    "    return genome\n",
    "\n",
    "#### challenge ####\n",
    "\n",
    "#f=open('../../Downloads/dataset_8224_2.txt','r')\n",
    "f=open('../../Downloads/rosalind_ba6k(2).txt','r')\n",
    "#f=open('input.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "genome=[map(int,lines[0][1:-1].split(' '))]\n",
    "\n",
    "i,j,k,l=map(int,lines[1].split(', '))\n",
    "\n",
    "\n",
    "result=two_break_on_genome(genome,i,j,k,l)\n",
    "final_output=[]\n",
    "output=[]\n",
    "for chrom in result:\n",
    "    output.append('('+' '.join([\"%+d\" % node for node in chrom])+')')\n",
    "final_output.append(' '.join(output))\n",
    "print '\\n'.join(final_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "'''\n",
    "ShortestRearrangementScenario(P, Q)\n",
    "     output P\n",
    "     RedEdges ← ColoredEdges(P)\n",
    "     BlueEdges ← ColoredEdges(Q)\n",
    "     BreakpointGraph ← the graph formed by RedEdges and BlueEdges\n",
    "     while BreakpointGraph has a non-trivial cycle Cycle\n",
    "          (j, i′) ← an arbitrary edge from BlueEdges in a nontrivial red-blue cycle\n",
    "          (i, j) ← an edge from RedEdges incident to node j\n",
    "          (i', j') ← an edge from RedEdges incident to node i'\n",
    "          RedEdges ← RedEdges with edges (i, j) and (i′, j′) removed\n",
    "          RedEdges ← RedEdges with edges (j, i′) and (j′, i) added\n",
    "          BreakpointGraph ← the graph formed by RedEdges and BlueEdges\n",
    "          P ← 2-BreakOnGenome(P, i, i′, j, j′)\n",
    "          output P\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, -2, -3, 4] [1, 2, -4, -3]\n",
      "[1, -2, -3, 4]\n",
      "1 4\n",
      "[[[2, 3], [4, 8], [7, 6], [5, 1], [2, 4], [3, 6], [5, 7], [8, 1]]]\n",
      "[['+1', -4, '+3', '+2']] None\n"
     ]
    }
   ],
   "source": [
    "def ColoredEdges(P):\n",
    "    Edges =[]\n",
    "    Chromosome=P\n",
    "    Nodes = [0]+ChromosomeToCycle(Chromosome)+[ChromosomeToCycle(Chromosome)[0]]\n",
    "    for j in range (1 , len(Chromosome)+1):\n",
    "        Edges.append([Nodes[2*j],Nodes[2*j+1]])\n",
    "    return Edges\n",
    "\n",
    "def ColoredEdges(Chromosome):\n",
    "    Edges =[]\n",
    "        #Chromosome=map(int,Chromosome.split(' '))\n",
    "    Nodes = [0]+ChromosomeToCycle(Chromosome)+[ChromosomeToCycle(Chromosome)[0]]\n",
    "    for j in range (1 , len(Chromosome)+1):\n",
    "        Edges.append([Nodes[2*j],Nodes[2*j+1]])\n",
    "    return Edges\n",
    "\n",
    "def TwoBreakOnGenomeGraph(GenomeGraph, i, ii, j, jj):\n",
    "    #print [i,ii]\n",
    "    if [i,ii] in GenomeGraph:\n",
    "        GenomeGraph.remove([i,ii])\n",
    "    else:\n",
    "        GenomeGraph.remove([ii,i])\n",
    "    if [j,jj] in GenomeGraph:\n",
    "        GenomeGraph.remove([j,jj])\n",
    "    else:\n",
    "        GenomeGraph.remove([jj,j])\n",
    "    \n",
    "    GenomeGraph.append([i,j])\n",
    "    GenomeGraph.append([ii,jj])\n",
    "    return GenomeGraph\n",
    "\n",
    "def TwoBreakOnGenome(P, i, ii, j, jj):\n",
    "    GenomeGraph = ColoredEdges(P) + BlackEdges(P)\n",
    "    #print GenomeGraph\n",
    "    GenomeGraph = sorted(TwoBreakOnGenomeGraph(GenomeGraph, i, ii, j, jj))\n",
    "    #print GenomeGraph\n",
    "    P=GraphToGenome(GenomeGraph)\n",
    "    return P\n",
    "def BreakpointGraph(Pgraph,Qgraph):\n",
    "    breakpointcycles=[]\n",
    "    breakpointpath=[]\n",
    "    #print Pgraph, Qgraph\n",
    "    while len(Pgraph) or len(Qgraph)>0:\n",
    "        i=0\n",
    "        cycle=[]\n",
    "        cycle+=Pgraph.pop(i)\n",
    "        graph='Qgraph'\n",
    "        while cycle[0] != cycle[-1]:\n",
    "            #graph='Qgraph'\n",
    "            if graph=='Qgraph':\n",
    "                for j in range(len(Qgraph)):\n",
    "                    edge=Qgraph[j]\n",
    "                    if cycle[-1] in edge:\n",
    "                        edge=Qgraph.pop(j)\n",
    "                        edge.remove(cycle[-1])\n",
    "                        cycle=cycle+edge\n",
    "                        graph='Pgraph'\n",
    "                        break\n",
    "            #print cycle\n",
    "            if graph=='Pgraph':\n",
    "                for j in range(len(Pgraph)):\n",
    "                    edge=Pgraph[j]\n",
    "                    if cycle[-1] in edge:\n",
    "                        edge=Pgraph.pop(j)\n",
    "                        edge.remove(cycle[-1])\n",
    "                        cycle=cycle+edge\n",
    "                        graph='Qgraph'\n",
    "                        break\n",
    "        breakpointcycles+=[cycle]\n",
    "    for cycle in breakpointcycles:\n",
    "        path=[]\n",
    "        for n in range(len(cycle)-1):\n",
    "            path+=[cycle[n:n+2]]\n",
    "            #print path\n",
    "        breakpointpath+=[path]    \n",
    "    return breakpointpath\n",
    "    \n",
    "\n",
    "def ShortestRearrangementScenario(P, Q):\n",
    "    print P\n",
    "    BlueEdges= ColoredEdges(Q)\n",
    "    RedEdges= ColoredEdges(P)\n",
    "    bpGraph=[BlueEdges+RedEdges]\n",
    "    #bpGraph=BreakpointGraph(Pgraph,Qgraph)\n",
    "    print len(bpGraph), len(P)\n",
    "    while len(bpGraph)<len(P):\n",
    "        RedEdges= ColoredEdges(P)\n",
    "        #print RedEdges\n",
    "        #print BlueEdges\n",
    "        #bpGraph= BreakpointGraph(RedEdges,BlueEdges)\n",
    "        #bpGraph=[BlueEdges+RedEdges]\n",
    "        #nonodes=len(Q)\n",
    "        print bpGraph\n",
    "        for cycle in bpGraph:\n",
    "            if len(cycle)>2:\n",
    "                [j,ii]=cycle[1]\n",
    "                j,ii=cycle[1][0],cycle[1][1]\n",
    "                i=[edge[0] for edge in cycle if edge[1]==j][0]\n",
    "                jj=[edge[1] for edge in cycle if edge[0]==ii][0]\n",
    "                #print j, ii, i, jj\n",
    "                RedEdges.remove([i,j])\n",
    "                RedEdges.remove([ii,jj])\n",
    "                RedEdges+=[j,ii]\n",
    "                RedEdges+=[jj,i]\n",
    "            bpGraph=[BlueEdges+RedEdges]\n",
    "            P = TwoBreakOnGenome(P, i, j, ii, jj)\n",
    "                #print cycle[1], cycle[3]\n",
    "            print P, \n",
    "    #While len(BreakpointGraph)<len nonodes:\n",
    "        \n",
    "\n",
    "        \n",
    "        \n",
    "### test 1 ###\n",
    "\n",
    "P='(+1 -2 -3 +4)'\n",
    "Q='(+1 +2 -4 -3)'\n",
    "\n",
    "\n",
    "#\n",
    "P=map(int,P[1:-1].split(' '))\n",
    "Q=map(int,Q[1:-1].split(' '))\n",
    "\n",
    "print P,Q\n",
    "print ShortestRearrangementScenario(P, Q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P,Q [[1, -2, -3, 4]] [[1, 2, -4, -3]]\n",
      "blueedge [[2, 3], [4, 8], [7, 6], [5, 1]] 4\n",
      "rededge [[2, 4], [3, 6], [5, 7], [8, 1]] 4 8\n",
      "blue-red cycles [[[2, 4], [4, 8], [8, 1], [1, 5], [5, 7], [7, 6], [6, 3], [3, 2]]] 1 4\n",
      "rededge [[2, 4], [3, 6], [5, 7], [8, 1]]\n",
      "cycle [3, 2]\n",
      "i,ii,j,jj 6 2 3 4\n",
      "rededge [[5, 7], [8, 1], [2, 6], [4, 3]]\n",
      "[[0, 3, 3, -4], [2, -2]]\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "bad operand type for unary -: 'list'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-44-8f999c47ca5e>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    244\u001b[0m \u001b[1;31m#Chromosome=[1, -2, -3, 4]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    245\u001b[0m \u001b[1;31m#print CycleToChromosome(ChromosomeToCycle(Chromosome))\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 246\u001b[1;33m \u001b[1;32mprint\u001b[0m \u001b[0mShortestRearrangementScenario\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mP\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mQ\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    247\u001b[0m \u001b[1;31m#i,ii,j,jj=4,8,2,1\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    248\u001b[0m \u001b[1;31m#print TwoBreakOnGenome(P, i, j, ii, jj)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-44-8f999c47ca5e>\u001b[0m in \u001b[0;36mShortestRearrangementScenario\u001b[1;34m(P, Q)\u001b[0m\n\u001b[0;32m    208\u001b[0m                 \u001b[1;32mprint\u001b[0m \u001b[0mGraphToGenome\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msorted\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mRedEdges\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mBkEdges\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    209\u001b[0m                 \u001b[0mre\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mGraphToGenome\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mRedEdges\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mBkEdges\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 210\u001b[1;33m                 \u001b[1;32mprint\u001b[0m \u001b[0mCycleToChromosome\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mre\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    211\u001b[0m                 \u001b[1;32mprint\u001b[0m \u001b[1;34m'rededge'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mRedEdges\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    212\u001b[0m                 \u001b[1;32mreturn\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-44-8f999c47ca5e>\u001b[0m in \u001b[0;36mCycleToChromosome\u001b[1;34m(Nodes)\u001b[0m\n\u001b[0;32m     25\u001b[0m             \u001b[0mChromosome\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mj\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mNodes\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0mj\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m/\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     26\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 27\u001b[1;33m             \u001b[0mChromosome\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mj\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0mNodes\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0mj\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m/\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     28\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mChromosome\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mkey\u001b[0m \u001b[1;32min\u001b[0m \u001b[0msorted\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mChromosome\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     29\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: bad operand type for unary -: 'list'"
     ]
    }
   ],
   "source": [
    "def ChromosomeToCycle(Chromosome):\n",
    "    '''input: Chromosome = [1, -2, -3, 4]\n",
    "        output: [1, 2, 4, 3, 6, 5, 7, 8] '''\n",
    "    Nodes={}\n",
    "    for j in range(len(Chromosome)):\n",
    "        i=Chromosome[j]\n",
    "        if i>0:\n",
    "            Nodes[2*j-1] = 2*i-1\n",
    "            Nodes[2*j] = 2*i\n",
    "        else:\n",
    "            Nodes[2*j-1] = -2*i\n",
    "            Nodes[2*j] = -2*i-1 \n",
    "    return [Nodes[key] for key in sorted(Nodes.keys())]\n",
    "\n",
    "def CycleToChromosome(Nodes):\n",
    "    '''Input: [1,2,4,3,6,5,7 8]\n",
    "        #output: ['+1', -2, -3, +4]\n",
    "        should be? [1,-2,-3,4]'''\n",
    "    Chromosome={}\n",
    "    Nodes=[0]+Nodes\n",
    "    #print Nodes\n",
    "    for j in range (1,(len(Nodes)/2)+1):\n",
    "        if Nodes[2*j-1] < Nodes[2*j]:\n",
    "            #Chromosome[j] = \"%+d\" % (Nodes[2*j]/2)\n",
    "            Chromosome[j] = Nodes[2*j]/2\n",
    "        else:\n",
    "            Chromosome[j] = - Nodes[2*j-1]/2\n",
    "    return [Chromosome[key] for key in sorted(Chromosome.keys())]\n",
    "\n",
    "def BlackEdges(P):\n",
    "    '''input: P=[[1,2,3,4],[5,6,7]]\n",
    "        output: [[1, 2], [3, 4], [5, 6], [7, 8], [9, 10], [11, 12], [13, 14]]\n",
    "    '''\n",
    "    Edges=[]\n",
    "    for Chromosome in P:\n",
    "        Chromosome=map(int,Chromosome)\n",
    "        ChrBlackEdges=[]\n",
    "        Nodes={}\n",
    "        for j in range(len(Chromosome)):\n",
    "            i=Chromosome[j]\n",
    "            if i>0:\n",
    "                Nodes[2*j-1] = 2*i-1\n",
    "                Nodes[2*j] = 2*i\n",
    "            else:\n",
    "                Nodes[2*j-1] = -2*i\n",
    "                Nodes[2*j] = -2*i-1 \n",
    "        theList=[Nodes[key] for key in sorted(Nodes.keys())]\n",
    "        N=2\n",
    "        ChrBlackEdges=[theList[n:n+N] for n in range(0, len(theList), N)]\n",
    "        Edges+=ChrBlackEdges\n",
    "    return Edges\n",
    "\n",
    "def ColoredEdges(P):\n",
    "    '''input: P=[[1,2,3,4],[5,6,7]]\n",
    "        output:[[2, 3], [4, 5], [6, 7], [8, 1], [10, 11], [12, 13], [14, 9]] '''\n",
    "    Edges =[]\n",
    "    for Chromosome in P:\n",
    "        #print 'coloredgechr',Chromosome\n",
    "        Nodes = [0]+ChromosomeToCycle(Chromosome)+[ChromosomeToCycle(Chromosome)[0]]\n",
    "        for j in range (1 , len(Chromosome)+1):\n",
    "            Edges.append([Nodes[2*j],Nodes[2*j+1]])\n",
    "    return Edges\n",
    "    \n",
    "def GraphToGenome(graph):\n",
    "    breakpointcycles=[]\n",
    "    ColoredEdges=[list(x) for x in graph]\n",
    "    while len(ColoredEdges) >0:\n",
    "        i=0\n",
    "        cycle=[]\n",
    "        cycle+=ColoredEdges.pop(i)\n",
    "        while cycle[0] != cycle[-1]:\n",
    "            for j in range(len(ColoredEdges)):\n",
    "                edge=ColoredEdges[j]\n",
    "                if cycle[-1] in edge:\n",
    "                    edge=ColoredEdges.pop(j)\n",
    "                    edge.remove(cycle[-1])\n",
    "                    cycle=cycle+edge\n",
    "                    break\n",
    "        cycle=cycle[:-1]\n",
    "        while cycle[-1]*2< cycle[-2]:\n",
    "            cycle=[cycle.pop(-1)]+cycle\n",
    "        #print cycle\n",
    "        breakpointcycles+=[[0]+cycle+[cycle[1]]]\n",
    "    return [CycleToChromosome(Chr) for Chr in breakpointcycles]\n",
    "\n",
    "\n",
    "'''def TwoBreakOnGenomeGraph(GenomeGraph, i, ii, j, jj):\n",
    "    #print [i,ii]\n",
    "    if [i,ii] in GenomeGraph:\n",
    "        GenomeGraph.remove([i,ii])\n",
    "    else:\n",
    "        GenomeGraph.remove([ii,i])\n",
    "    if [j,jj] in GenomeGraph:\n",
    "        GenomeGraph.remove([j,jj])\n",
    "    else:\n",
    "        GenomeGraph.remove([jj,j])\n",
    "    \n",
    "    GenomeGraph.append([i,j])\n",
    "    GenomeGraph.append([ii,jj])\n",
    "    return GenomeGraph'''\n",
    "\n",
    "'''def TwoBreakOnGenome(P, i, ii, j, jj):\n",
    "    GenomeGraph = ColoredEdges(P) + BlackEdges(P)\n",
    "    #print GenomeGraph\n",
    "    GenomeGraph = sorted(TwoBreakOnGenomeGraph(GenomeGraph, i, ii, j, jj))\n",
    "    #print GenomeGraph\n",
    "    P=GraphToGenome(GenomeGraph)\n",
    "    return P'''\n",
    "\n",
    "def BreakpointGraph(RedEdges,BlueEdges):\n",
    "    '''input:\n",
    "        Pgraph= [[2, 4], [3, 6], [5, 7], [8, 1]]\n",
    "        Qgraph= [[2, 3], [4, 8], [7, 6], [5, 1]]\n",
    "    output:\n",
    "        [[[2, 4], [4, 8], [8, 1], [1, 5], [5, 7], [7, 6], [6, 3], [3, 2]]]\n",
    "    '''\n",
    "    Pgraph=[list(x) for x in RedEdges]\n",
    "    Qgraph=[list(x) for x in BlueEdges]\n",
    "    breakpointcycles=[]\n",
    "    breakpointpath=[]\n",
    "    #print Pgraph, Qgraph\n",
    "    while len(Pgraph) or len(Qgraph)>0:\n",
    "        i=0\n",
    "        cycle=[]\n",
    "        cycle+=Pgraph.pop(i)\n",
    "        graph='Qgraph'\n",
    "        while cycle[0] != cycle[-1]:\n",
    "            #graph='Qgraph'\n",
    "            if graph=='Qgraph':\n",
    "                for j in range(len(Qgraph)):\n",
    "                    edge=Qgraph[j]\n",
    "                    if cycle[-1] in edge:\n",
    "                        edge=Qgraph.pop(j)\n",
    "                        edge.remove(cycle[-1])\n",
    "                        cycle=cycle+edge\n",
    "                        graph='Pgraph'\n",
    "                        break\n",
    "            #print cycle\n",
    "            if graph=='Pgraph':\n",
    "                for j in range(len(Pgraph)):\n",
    "                    edge=Pgraph[j]\n",
    "                    if cycle[-1] in edge:\n",
    "                        edge=Pgraph.pop(j)\n",
    "                        edge.remove(cycle[-1])\n",
    "                        cycle=cycle+edge\n",
    "                        graph='Qgraph'\n",
    "                        break\n",
    "        breakpointcycles+=[cycle]\n",
    "    for cycle in breakpointcycles:\n",
    "        path=[]\n",
    "        for n in range(len(cycle)-1):\n",
    "            path+=[cycle[n:n+2]]\n",
    "            #print path\n",
    "        breakpointpath+=[path]    \n",
    "    return breakpointpath\n",
    "    \n",
    "\n",
    "def ShortestRearrangementScenario(P, Q):\n",
    "    print 'P,Q', P,Q\n",
    "    BlueEdges = ColoredEdges(Q)\n",
    "    RedEdges =ColoredEdges(P)\n",
    "    BkEdges= BlackEdges(P)\n",
    "    #print 'blueedge',BlueEdges\n",
    "    #print 'rededge',RedEdges\n",
    "    #bpGraph=[BlueEdges+RedEdges]\n",
    "    bpGraphcycles=BreakpointGraph(RedEdges,BlueEdges)\n",
    "    print 'blueedge',BlueEdges, len(BlueEdges)\n",
    "    print 'rededge',RedEdges, len(RedEdges), len([item for sublist in P for item in sublist])*2\n",
    "    print 'blue-red cycles',bpGraphcycles, len(bpGraphcycles),len([item for sublist in P for item in sublist])\n",
    "    #print 'bpGraph ', bpGraph\n",
    "    #print len(bpGraph), len(P)\n",
    "    while len(bpGraphcycles)<len([item for sublist in P for item in sublist]):\n",
    "    #while RedEdges!= BlackEdges(P):\n",
    "        #RedEdges =ColoredEdges(P)\n",
    "        #bpGraphcycles=BreakpointGraph(RedEdges,BlueEdges)\n",
    "        for cycle in bpGraphcycles:\n",
    "            if len(cycle)<=2:\n",
    "                pass\n",
    "            else:\n",
    "                print 'rededge',RedEdges\n",
    "                print 'cycle',cycle[-1]\n",
    "                j,ii=cycle[-1][0],cycle[-1][1]\n",
    "                ivalue=[edge[:] for edge in RedEdges if j in edge][0]\n",
    "                #print 'rededge',RedEdges\n",
    "                ivalue.remove(j)\n",
    "                #print 'rededge',RedEdges\n",
    "                i=ivalue[0]\n",
    "                #print 'rededge',RedEdges\n",
    "                #print i\n",
    "                jjvalue=[edge[:] for edge in RedEdges if ii in edge][0]\n",
    "                jjvalue.remove(ii)\n",
    "                jj=jjvalue[0]\n",
    "                print 'i,ii,j,jj',i, ii, j ,jj\n",
    "                #print RedEdges\n",
    "                #print RedEdges.index([j,i])\n",
    "                try:\n",
    "                    RedEdges.remove([i,j])\n",
    "                except ValueError:\n",
    "                    RedEdges.remove([j,i])\n",
    "                try:\n",
    "                    RedEdges.remove([ii,jj])\n",
    "                except ValueError:\n",
    "                    RedEdges.remove([jj,ii])\n",
    "                RedEdges+=[[ii,i]]\n",
    "                RedEdges+=[[jj,j]]\n",
    "                \n",
    "                print 'rededge',RedEdges\n",
    "                print GraphToGenome(sorted(RedEdges)+BkEdges)\n",
    "                re=GraphToGenome(RedEdges+BkEdges)\n",
    "                print CycleToChromosome(re)\n",
    "                print 'rededge',RedEdges\n",
    "                return\n",
    "        \n",
    "        '''#re= GraphToGenome(sorted(RedEdges[:]))\n",
    "        #print re\n",
    "        #print CycleToChromosome(re)\n",
    "        #print 'RedEdges',re\n",
    "        bpGraphcycles=BreakpointGraph(RedEdges,BlueEdges)\n",
    "        print bpGraphcycles\n",
    "        #bpGraphcycles=BreakpointGraph([re],[be])\n",
    "        #P = [item for item in TwoBreakOnGenomeGraph(RedEdges, i, j, ii, jj)]\n",
    "        #print 'P',P\n",
    "        #P= TwoBreakOnGenome(P, i, j, ii, jj)\n",
    "                #print cycle[1], cycle[3]\n",
    "        #print P\n",
    "        #bpGraphcycles=BreakpointGraph(RedEdges,BlueEdges)\n",
    "        #print 'bpcycles',bpGraphcycles\n",
    "    return\n",
    "    #While len(BreakpointGraph)<len nonodes:\n",
    "        '''\n",
    "\n",
    "        \n",
    "        \n",
    "### test 1 ###\n",
    "\n",
    "P='(+1 -2 -3 +4)'\n",
    "Q='(+1 +2 -4 -3)'\n",
    "\n",
    "\n",
    "#\n",
    "P=[map(int,P[1:-1].split(' '))]\n",
    "Q=[map(int,Q[1:-1].split(' '))]\n",
    "\n",
    "#Chromosome=[1, -2, -3, 4]\n",
    "#print CycleToChromosome(ChromosomeToCycle(Chromosome))\n",
    "print ShortestRearrangementScenario(P, Q)\n",
    "#i,ii,j,jj=4,8,2,1\n",
    "#print TwoBreakOnGenome(P, i, j, ii, jj)\n",
    "#Pgraph= [[2, 4], [3, 6], [5, 7], [8, 1]]\n",
    "#Qgraph= [[2, 3], [4, 8], [7, 6], [5, 1]]\n",
    "\n",
    "#print BreakpointGraph(Pgraph,Qgraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'return' outside function (<ipython-input-22-d60924a8a1c8>, line 20)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-22-d60924a8a1c8>\"\u001b[1;36m, line \u001b[1;32m20\u001b[0m\n\u001b[1;33m    return path\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m 'return' outside function\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1, -2, -3, 4]] [[1, 2, -4, -3]]\n",
      "queue set([])\n",
      "current 1 graph [2, -2] remaining set([2, 3, 4, -2, -4, -3, -1])\n",
      "set([2, -2])\n",
      "queue set([-2])\n",
      "current 2 graph [1, 4] remaining set([3, 4, -4, -3, -1])\n",
      "set([4])\n",
      "queue set([-2])\n",
      "current 4 graph [-1, 2] remaining set([3, -4, -3, -1])\n",
      "set([-1])\n",
      "queue set([-2])\n",
      "current -1 graph [4, -3] remaining set([3, -4, -3])\n",
      "set([-3])\n",
      "queue set([-2])\n",
      "current -3 graph [-4, -1] remaining set([3, -4])\n",
      "set([-4])\n",
      "queue set([-4])\n",
      "current -2 graph [3, 1] remaining set([3])\n",
      "set([3])\n",
      "queue set([-4])\n",
      "current 3 graph [-2, -4] remaining set([])\n",
      "set([])\n",
      "queue set([])\n",
      "current -4 graph [-3, 3] remaining set([])\n",
      "set([])\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "\n",
    "'''\n",
    "A solution to a code challenges that accompanies Bioinformatics Algorithms: An Active-nearning Approach by Phillip Compeau & Pavel Pevzner.\n",
    "The textbook is hosted on Stepic and the problem is listed on ROSALIND under the Textbook Track.\n",
    "Problem Title: 2-Break Distance Problem\n",
    "Rosalind ID: 6C\n",
    "URL: http://rosalind.info/problems/6c/\n",
    "'''\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "def two_break_dist(P, Q):\n",
    "    '''Returns the 2-Break Distance of Circular Chromosomes P and Q.'''\n",
    "\n",
    "    # Construct the break point graph of P and Q.\n",
    "    graph = defaultdict(list)\n",
    "    for perm_cycle in P+Q:\n",
    "        n = len(perm_cycle)\n",
    "        for i in xrange(n):\n",
    "            # Add the edge between consecutive items (both orders since the breakpoint graph is undirected).\n",
    "            # Note: Modulo n in the higher index for the edge between the last and first elements.\n",
    "            graph[perm_cycle[i]].append(-1*perm_cycle[(i+1) % n])\n",
    "            graph[-1*perm_cycle[(i+1) % n]].append(perm_cycle[i])\n",
    "    #print graph\n",
    "    # Traverse the breakpoint graph to get the number of connected components.\n",
    "    component_count = 0\n",
    "    remaining = set(graph.keys())\n",
    "    while remaining:\n",
    "        component_count += 1\n",
    "        queue = {remaining.pop()}  # Undirected graph, so we can choose a remaining node arbitrarily.\n",
    "        while queue:\n",
    "            # Select an element from the queue and get its remaining children.\n",
    "            current = queue.pop()\n",
    "            #print 'queue',queue\n",
    "            #print 'current',current, 'graph',graph[current], 'remaining',remaining\n",
    "            new_nodes = {node for node in graph[current] if node in remaining}\n",
    "            #print new_nodes\n",
    "            # Add the new nodes to the queue, remove them from the remaining nodes.\n",
    "            queue |= new_nodes\n",
    "            remaining -= new_nodes\n",
    "\n",
    "    # Theorem: d(P,Q) = blocks(P,Q) - cycles(P,Q)\n",
    "    return sum(map(len,P)) - component_count\n",
    "\n",
    "\n",
    "lines='(+1 -2 -3 +4)','(+1 +2 -4 -3)'\n",
    "P, Q = [line.strip().lstrip('(').rstrip(')').split(')(') for line in lines]\n",
    "\n",
    "#\n",
    "\n",
    "P = [map(int, perm_cycle.split()) for perm_cycle in P]\n",
    "Q = [map(int, perm_cycle.split()) for perm_cycle in Q]\n",
    "\n",
    "#print P,Q\n",
    "print two_break_dist(P, Q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['(-8 +2 -5 -13 +3 +6 +12 -9 +7 +4 +11 +1 +10)', '(-5 +13 -2 -12 -10 +1 -4 -8 -3 +6 +11 -7 +9)']\n",
      "(-8 +2 -5 -13 +3 +6 +12 -9 +7 +4 +11 +1 +10)\n",
      "(-8 +2 -5 -13 +3 +6 +12 -9 +7 -11 -4 +1 +10)\n",
      "(-8 +2 -5 -13 +3 +6 +12 -9 +7 -11 -4) (+1 +10)\n",
      "(-8 +2 -5 -13 +3 +6 +1 +10 +12 -9 +7 -11 -4)\n",
      "(-8 +2 -5 -13 +3 +6 +11 -7 +9 -12 -10 -1 -4)\n",
      "(-8 +2 -5 -13 +3 +6 +11 -7 +9 -12 -10 +1 -4)\n",
      "(+2 -5 -13 +3 +8 +4 -1 +10 +12 -9 +7 -11 -6)\n",
      "(+2 -5 -13) (+3 +8 +4 -1 +10 +12 -9 +7 -11 -6)\n",
      "(+2 -5 +13) (+3 +8 +4 -1 +10 +12 -9 +7 -11 -6)\n",
      "(-5 +13 -2) (+3 +8 +4 -1 +10 +12 -9 +7 -11 -6)\n",
      "(-5 +13 -2 -12 -10 +1 -4 -8 -3 +6 +11 -7 +9)\n"
     ]
    }
   ],
   "source": [
    "def chromosome_to_cycle(p):\n",
    "    '''\n",
    "    CODE CHALLENGE: Implement ChromosomeToCycle.\n",
    "    Input: A chromosome Chromosome containing n synteny blocks.\n",
    "    Output: The sequence Nodes of integers between 1 and 2n resulting \n",
    "    from applying ChromosomeToCycle to Chromosome.\n",
    "    '''\n",
    "    nodes = []\n",
    "    \n",
    "    for i in p:\n",
    "        if (i>0):\n",
    "            nodes.append(2*i-1)\n",
    "            nodes.append(2*i)\n",
    "        else:\n",
    "            nodes.append(-2*i)\n",
    "            nodes.append(-2*i-1)\n",
    "    return nodes\n",
    "\n",
    "def cycle_to_chromosome(nodes):\n",
    "    '''\n",
    "    CODE CHALLENGE: Implement CycleToChromosome.\n",
    "    Input: A sequence Nodes of integers between 1 and 2n.\n",
    "    Output: The chromosome Chromosome containing n synteny blocks resulting \n",
    "    from applying CycleToChromosome to Nodes.\n",
    "    '''\n",
    "    p = []\n",
    "    for j in range(0,len(nodes)/2):\n",
    "        if nodes[2*j] < nodes[2*j+1]:\n",
    "            s = j+1\n",
    "        else:\n",
    "            s = -(j+1)\n",
    "        p.append(s)\n",
    "    return p\n",
    "\n",
    "def colored_edges(genome):\n",
    "    '''\n",
    "    CODE CHALLENGE: Implement ColoredEdges.\n",
    "    Input: A genome P.\n",
    "    Output: The collection of colored edges in the genome graph of P \n",
    "    in the form (x, y).\n",
    "    '''\n",
    "    g = []\n",
    "    for p in genome:\n",
    "        s = chromosome_to_cycle(p)\n",
    "        for j in range(len(s)/2):\n",
    "            head = 1+2*j\n",
    "            tail = (2+2*j) % len(s)\n",
    "            e = (s[head],s[tail])\n",
    "            g.append(e)\n",
    "    return g\n",
    "\n",
    "def graph_to_genome(g):\n",
    "    '''\n",
    "    CODE CHALLENGE: Implement GraphToGenome.\n",
    "    Input: The colored edges ColoredEdges of a genome graph.\n",
    "    Output: The genome P corresponding to this genome graph.\n",
    "    '''\n",
    "    import numpy as np\n",
    "    genome = []\n",
    "    visit = []\n",
    "    adj = np.zeros(len(g)*2, dtype = np.int)\n",
    "    for t in g:\n",
    "        adj[t[0]-1] = t[1]-1\n",
    "        adj[t[1]-1] = t[0]-1\n",
    "    \n",
    "    for t in g:\n",
    "        orig = t[0]\n",
    "        if orig in visit:\n",
    "            continue\n",
    "        visit.append(orig)\n",
    "        if (orig%2 == 0):\n",
    "            closing = orig-1\n",
    "        else:\n",
    "            closing = orig+1\n",
    "        p = []\n",
    "        i = 0\n",
    "        while(True):\n",
    "            if (orig%2 == 0):\n",
    "                p.append(orig/2)\n",
    "            else:\n",
    "                p.append(-(orig+1)/2)\n",
    "            dest = adj[orig-1]+1\n",
    "            i = i + 1\n",
    "            if (i>100):\n",
    "                assert False\n",
    "                return\n",
    "            visit.append(dest)\n",
    "            if (dest == closing):\n",
    "                genome.append(p)\n",
    "                break\n",
    "            if (dest%2 == 0):\n",
    "                orig = dest -1\n",
    "            else:\n",
    "                orig = dest + 1\n",
    "            assert orig > 0\n",
    "            visit.append(orig)\n",
    "    return genome\n",
    "\n",
    "\n",
    "\n",
    "def two_break_dist(P, Q):\n",
    "    '''Returns the 2-Break Distance of Circular Chromosomes P and Q.'''\n",
    "\n",
    "    graph = defaultdict(list)\n",
    "    for perm_cycle in P+Q:\n",
    "        n = len(perm_cycle)\n",
    "        for i in xrange(n):\n",
    "            graph[perm_cycle[i]].append(-1*perm_cycle[(i+1) % n])\n",
    "            graph[-1*perm_cycle[(i+1) % n]].append(perm_cycle[i])\n",
    "    component_count = 0\n",
    "    remaining = set(graph.keys())\n",
    "    while remaining:\n",
    "        component_count += 1\n",
    "        queue = {remaining.pop()}  # Undirected graph, so we can choose a remaining node arbitrarily.\n",
    "        while queue:\n",
    "            current = queue.pop()\n",
    "            new_nodes = {node for node in graph[current] if node in remaining}\n",
    "            queue |= new_nodes\n",
    "            remaining -= new_nodes\n",
    "    return sum(map(len,P)) - component_count\n",
    "\n",
    "def colored_edges_cycles(blue, red):\n",
    "    '''    \n",
    "    returns all alternating red-blue-edge cycles\n",
    "    '''\n",
    "    import numpy as np\n",
    "    cycles = []\n",
    "    size = len(blue)+len(red) \n",
    "    adj = np.zeros(shape = (size,2), dtype = np.int)\n",
    "    visited = np.zeros(shape = size, dtype = np.bool)\n",
    "    for e in blue:\n",
    "        adj[e[0]-1,0] = e[1]-1\n",
    "        adj[e[1]-1,0] = e[0]-1\n",
    "    for e in red:\n",
    "        adj[e[0]-1,1] = e[1]-1\n",
    "        adj[e[1]-1,1] = e[0]-1\n",
    "    \n",
    "    for node in range(size):\n",
    "        if not visited[node]:\n",
    "            visited[node] = True\n",
    "            head = node\n",
    "            cycle = [head+1]\n",
    "            # arbitrary we start with a blue edge\n",
    "            color = 0\n",
    "            while (True):\n",
    "                node = adj[node,color]\n",
    "                if (node == head):\n",
    "                    # cycle found, we got back to the visited head node, \n",
    "                    # just break\n",
    "                    cycles.append(cycle)\n",
    "                    break\n",
    "                cycle.append(node+1)\n",
    "                visited[node] = True\n",
    "                color = (color+1) % 2\n",
    "    return cycles\n",
    "\n",
    "def two_break_on_genome_graph(g,i,j,k,l):\n",
    "    '''\n",
    "    CODE CHALLENGE: Implement 2-BreakOnGenomeGraph.\n",
    "    Input: The colored edges of a genome graph GenomeGraph, \n",
    "    followed by indices i, j, k, and l.\n",
    "    Output: The colored edges of the genome graph resulting from applying \n",
    "    the 2-break operation 2-BreakOnGenomeGraph(GenomeGraph, i, j, k, l).\n",
    "    '''\n",
    "    bg = []\n",
    "    # equivalent and more elegant but not accepted by the grader ...\n",
    "#    d = {(i,j):(i,k), (j,i):(j,l), (k,l):(k,i), (l,k):(l,j)}    \n",
    "#    for t in g:\n",
    "#        if (t in d):\n",
    "#            bg.append(d[t])\n",
    "#        else:\n",
    "#            bg.append(t)\n",
    "    \n",
    "    # so do it this way\n",
    "    rem = ((i,j), (j,i), (k,l), (l,k))\n",
    "    bg = [ t for t in g if t not in rem]\n",
    "    bg.append((i,k))\n",
    "    bg.append((j,l))\n",
    "    \n",
    "    return bg\n",
    "\n",
    "def two_break_on_genome(genome,i,j,k,l):\n",
    "    '''\n",
    "    CODE CHALLENGE: Implement 2-BreakOnGenome.\n",
    "    Input: A genome P, followed by indices i, i', j, and j'.\n",
    "    Output: The genome P' resulting from applying the 2-break operation\n",
    "    2-BreakOnGenomeGraph(GenomeGraph, i, i′, j, j′).\n",
    "    '''\n",
    "    g = colored_edges(genome)\n",
    "    g = two_break_on_genome_graph(g,i,j,k,l)\n",
    "    genome = graph_to_genome(g)\n",
    "    return genome\n",
    "\n",
    "def two_break_sorting(P,Q):\n",
    "    '''\n",
    "    CODE CHALLENGE: Solve the 2-Break Sorting Problem.     \n",
    "    2-Break Sorting Problem: Find a shortest transformation \n",
    "    of one genome into another via 2-breaks.\n",
    "    Input: Two genomes with circular chromosomes on the same \n",
    "    set of synteny blocks.\n",
    "    Output: The collection of genomes resulting from applying \n",
    "    a shortest sequence of 2-breaks transforming one genome into the other.\n",
    "    '''\n",
    "    red = colored_edges(Q)\n",
    "    path = [P]\n",
    "    while two_break_dist(P, Q) > 0:\n",
    "        cycles = colored_edges_cycles(colored_edges(P),red)\n",
    "        for c in cycles:\n",
    "            if len(c) >= 4:\n",
    "                P = two_break_on_genome(P,c[0],c[1],c[3],c[2])\n",
    "                path.append(P)\n",
    "                break          \n",
    "    return path\n",
    "\n",
    "###test1 ###\n",
    "\n",
    "lines='(+1 -2 -3 +4)','(+1 +2 -4 -3)'\n",
    "P, Q = [line.strip().lstrip('(').rstrip(')').split(')(') for line in lines]\n",
    "\n",
    "\n",
    "### test 2 ###\n",
    "f=open('input.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "#print lines\n",
    "P, Q = [line.strip().lstrip('(').rstrip(')').split(')(') for line in lines]\n",
    "\n",
    "\n",
    "### challenge : WORKED###\n",
    "\n",
    "#f=open('../../Downloads/dataset_288_5(1).txt','r')\n",
    "f=open('../../Downloads/rosalind_ba6d.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "print lines\n",
    "P, Q = [line.strip().lstrip('(').rstrip(')').split(')(') for line in lines]\n",
    "\n",
    "#\n",
    "\n",
    "\n",
    "\n",
    "P = [map(int, perm_cycle.split()) for perm_cycle in P]\n",
    "Q = [map(int, perm_cycle.split()) for perm_cycle in Q]\n",
    "#print P,Q\n",
    "\n",
    "result=two_break_sorting(P,Q)\n",
    "final_output=[]\n",
    "for step in result:\n",
    "    output=[]\n",
    "    for chrom in step:\n",
    "        output.append('('+' '.join([\"%+d\" % node for node in chrom])+')')\n",
    "    final_output.append(' '.join(output))\n",
    "print '\\n'.join(final_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
