{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': ['0'], '0': ['3'], '3': ['2'], '2': ['1', '6'], '5': ['4'], '4': ['2'], '7': ['9'], '6': ['5', '8'], '9': ['6'], '8': ['7']}\n",
      "2->6->8->7->9->6->5->4->2->1->0->3->2\n"
     ]
    }
   ],
   "source": [
    "'''     EULERIANCYCLE(Graph)\n",
    "        form a cycle Cycle by randomly walking in Graph (don't visit the same edge twice!)\n",
    "        while there are unexplored edges in Graph\n",
    "            select a node newStart in Cycle with still unexplored edges\n",
    "            form Cycle’ by traversing Cycle (starting at newStart) and then randomly walking\n",
    "            Cycle ← Cycle’\n",
    "        return Cycle '''\n",
    "\n",
    "\n",
    "''' RAW '''\n",
    "\n",
    "from collections import defaultdict\n",
    "import re\n",
    "import random\n",
    "import copy\n",
    "from itertools import cycle, islice, dropwhile\n",
    "\n",
    "lines= ['0 -> 3', '1 -> 0', '2 -> 1,6', '3 -> 2', '4 -> 2', '5 -> 4', '6 -> 5,8', '7 -> 9', '8 -> 7', '9 -> 6']\n",
    "\n",
    "Graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    Graph[data[0]]=data[1:]\n",
    "print Graph \n",
    "\n",
    "\n",
    "''' maintain the current cycle that Leo is building x\n",
    "the list of unused edges incident to each node \n",
    "the list of nodes on the current cycle that have unused edges.'''\n",
    " #dic with key-prefix,values=suffix\n",
    "# def EULERIANCYCLE(Graph):\n",
    "\n",
    "# 1. form a cycle Cycle by randomly walking in Graph (don't visit the same edge twice!)\n",
    "\n",
    "graph_unused=copy.deepcopy(Graph)\n",
    "cycle_Leo=[]\n",
    "step=1\n",
    "Start=random.choice(graph_unused.keys()) #Pick a start position\n",
    "Position=Start #Move Leo to start position\n",
    "cycle_Leo.append(Position) #record Leo position\n",
    "\n",
    "while graph_unused[Position]: # If next move possible\n",
    "    step+=1\n",
    "    NextStep=random.choice(graph_unused[Position]) #Pick a place to move Leo\n",
    "    graph_unused[Position].remove(NextStep) # Remove this place from list of available positions to move to\n",
    "    Position=NextStep # Move Leo to position\n",
    "    cycle_Leo.append(Position) # Record Leo position\n",
    "\n",
    "#print 'Step '+str(step)+'\\ngraph_unused: '+str(graph_unused)+'\\ncycle_Leo: '+str(cycle_Leo)\n",
    "\n",
    "# 2. while there are unexplored edges in Graph, select a node newStart in Cycle with still unexplored edges,form Cycle’ by traversing Cycle (starting at newStart) and then randomly walking select a node newStart in Cycle with still unexplored edges\n",
    "\n",
    "while bool([a for a in graph_unused.values() if a != []]) is True: # while there are unexplored edges in Graph\n",
    "    #select a node newStart in Cycle with still unexplored edges\n",
    "    for edge in cycle_Leo:\n",
    "        if graph_unused[edge]:\n",
    "            NewStart=edge\n",
    "            print NewStart\n",
    "            break\n",
    "    #form Cycle’ by traversing Cycle (starting at newStart)...\n",
    "        #  This is done by shifting cycle_Leo so it starts and ends at NewStart\n",
    "    cycled= cycle(cycle_Leo[1:])  # cycle through cycle_Leo (removing duplicated start position)\n",
    "    skipped = dropwhile(lambda x: x != NewStart, cycled)  # drop the values until x==NewStart\n",
    "    sliced = islice(skipped, None, len(cycle_Leo))  # take the first n values\n",
    "    cycle_Leo = list(sliced)  # create a list from iterator\n",
    "    Position=NewStart\n",
    "    print Position\n",
    "    #...and then randomly walking Cycle\n",
    "    while graph_unused[Position]: #if next move possible\n",
    "        NextStep=random.choice(graph_unused[Position]) #Pick a place to move\n",
    "        graph_unused[Position].remove(NextStep) #remove this place from list of movable positions\n",
    "        Position=NextStep # make the move\n",
    "        cycle_Leo.append(Position) #record the move\n",
    "    #print cycle_Leo\n",
    "fa=open('answer.txt','w')\n",
    "entry='->'.join(cycle_Leo)\n",
    "#print entry\n",
    "fa.write(entry)\n",
    "fa.close()\n",
    "\n",
    "\n",
    "''' FUNCTION '''\n",
    "\n",
    "def EULERIANCYCLE(Graph):\n",
    "    from collections import defaultdict\n",
    "    import re\n",
    "    import random\n",
    "    import copy\n",
    "    from itertools import cycle, islice, dropwhile\n",
    "\n",
    "    # 1. form a cycle Cycle by randomly walking in Graph (don't visit the same edge twice!)\n",
    "\n",
    "    graph_unused=copy.deepcopy(Graph)\n",
    "    cycle_Leo=[]\n",
    "    Start=random.choice(graph_unused.keys()) #Pick a start position\n",
    "    Position=Start #Move Leo to start position\n",
    "    cycle_Leo.append(Position) #record Leo position\n",
    "\n",
    "    while graph_unused[Position]: # If next move possible\n",
    "        NextStep=random.choice(graph_unused[Position]) #Pick a place to move Leo\n",
    "        graph_unused[Position].remove(NextStep) # Remove this place from list of available positions to move to\n",
    "        Position=NextStep # Move Leo to position\n",
    "        cycle_Leo.append(Position) # Record Leo position\n",
    "\n",
    "    # 2. while there are unexplored edges in Graph, select a node newStart in Cycle with still unexplored edges,form Cycle’ by traversing Cycle (starting at newStart) and then randomly walking select a node newStart in Cycle with still unexplored edges\n",
    "\n",
    "    while bool([a for a in graph_unused.values() if a != []]) is True: # while there are unexplored edges in Graph\n",
    "        #select a node newStart in Cycle with still unexplored edges\n",
    "        for edge in cycle_Leo:\n",
    "            if graph_unused[edge]:\n",
    "                NewStart=edge\n",
    "                break\n",
    "        #form Cycle’ by traversing Cycle (starting at newStart)...\n",
    "            #  This is done by shifting cycle_Leo so it starts and ends at NewStart\n",
    "        cycled= cycle(cycle_Leo[1:])  # cycle through cycle_Leo (removing duplicated start position)\n",
    "        skipped = dropwhile(lambda x: x != NewStart, cycled)  # drop the values until x==NewStart\n",
    "        sliced = islice(skipped, None, len(cycle_Leo))  # take the first n values\n",
    "        cycle_Leo = list(sliced)  # create a list from iterator\n",
    "        Position=NewStart\n",
    "        #...and then randomly walking Cycle\n",
    "        while graph_unused[Position]: #if next move possible\n",
    "            NextStep=random.choice(graph_unused[Position]) #Pick a place to move\n",
    "            graph_unused[Position].remove(NextStep) #remove this place from list of movable positions\n",
    "            Position=NextStep # make the move\n",
    "            cycle_Leo.append(Position) #record the move\n",
    "    return cycle_Leo \n",
    "\n",
    "\n",
    "print '->'.join(EULERIANCYCLE(Graph))\n",
    "\n",
    "\n",
    "''' CHALLENGE '''\n",
    "\n",
    "f=open('../../Downloads/dataset_203_2.txt','r')\n",
    "#f=open('input.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "Graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    Graph[data[0]]=data[1:]\n",
    "#print Graph \n",
    "\n",
    "fa=open('answer.txt','w')\n",
    "fa.write('->'.join(EULERIANCYCLE(Graph)))\n",
    "fa.close()\n",
    "\n",
    "''' ROSALIND SOLUTION '''\n",
    "''' when random, I can just pick the first in list. pop() function returns the removed object from the list.'''\n",
    "def eulerianCycle(graph):\n",
    "    # init cycle with any node\n",
    "    cycle = [graph.keys()[0]]\n",
    "    # run until all edges are moved from graph to cycle\n",
    "    while len(graph) > 0:\n",
    "        # whenever cycle closes rotate to a node with remaining targets\n",
    "        if cycle[0] == cycle[-1]:\n",
    "            while not cycle[0] in graph:\n",
    "                cycle.pop(0)\n",
    "                cycle.append(cycle[0])\n",
    "        # the last node of cycle is the new source\n",
    "        source = cycle[-1]\n",
    "        # move one target at a time from graph to the end of cycle\n",
    "        cycle.append(graph[source].pop())\n",
    "        # clean up empty dict entries of graph\n",
    "        if len(graph[source]) == 0: del graph[source]\n",
    "    return cycle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 6\n",
      "4->['6']\n",
      "['1', '3', '4', '6', '7', '8', '9', '6', '3', '0', '2', '1']\n",
      "['3', '4', '6', '7', '8', '9', '6', '3', '0', '2', '1']\n",
      "['6', '7', '8', '9', '6', '3', '0', '2', '1', '3', '4']\n"
     ]
    }
   ],
   "source": [
    "''' Solve the Eulerian Path Problem.'''\n",
    "\n",
    "''' RAW \n",
    "Doesn't work all the time; pb at then end when I need to break up the loop. I cant break it up at the right place'''\n",
    "import re\n",
    "#from Bioinf_functions import eul_cycle\n",
    "f=open('../../Downloads/dataset_203_5.txt','r')\n",
    "#lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "lines=['0 -> 2','1 -> 3','2 -> 1','3 -> 0,4','6 -> 3,7','7 -> 8','8 -> 9','9 -> 6']\n",
    "Graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    Graph[data[0]]=data[1:]\n",
    "\n",
    "#Find unbalanced nodes\n",
    "from collections import Counter\n",
    "from collections import defaultdict\n",
    "#print Graph\n",
    "#print Counter(value for values in Graph.itervalues() for value in values)\n",
    "#numout=dict([(k,(len(Graph[k]))) for k in Graph])\n",
    "#outdegrees=Counter(k,len(v) for k,v in Graph.keys())\n",
    "outdegrees=Counter(dict([(k,(len(Graph[k]))) for k in Graph]))\n",
    "indegrees=Counter(value for values in Graph.itervalues() for value in values)\n",
    "#print indegrees, outdegrees\n",
    "\n",
    "#print [k for k in outdegrees+indegrees]\n",
    "for k in outdegrees+indegrees:\n",
    "    if outdegrees[k]<indegrees[k]:\n",
    "        #print k, outdegrees[k],indegrees[k]\n",
    "        pre=k\n",
    "        \n",
    "    if outdegrees[k]>indegrees[k]:\n",
    "        #print k, outdegrees[k],indegrees[k]\n",
    "        post=k\n",
    "print pre,post\n",
    "if pre in Graph:\n",
    "    Graph[pre].append(post)\n",
    "else:\n",
    "    Graph[pre]=[post]\n",
    "print str(pre)+'->'+str(Graph[pre])\n",
    "#print Graph.keys()\n",
    "#print Graph\n",
    "# find eulerian cycle\n",
    "eul_cycle=eulerianCycle(Graph)\n",
    "print eul_cycle\n",
    "print eul_cycle[1:]\n",
    "#print eul_cycle\n",
    "#cut cycle at fake link\n",
    "#cycled= cycle(eul_cycle[1:])  # cycle through cycle_Leo (removing duplicated start position)\n",
    "#skipped = dropwhile(lambda x: x != pre, cycled)  # drop the values until x==NewStart\n",
    "#sliced = islice(skipped, None, len(eul_cycle)-1)  # take the first n values\n",
    "#eul_path = list(sliced)  # create a list from iterator\n",
    "eul_cycle=eul_cycle[1:]\n",
    "while eul_cycle[0] != post or eul_cycle[-1] != pre:  \n",
    "    eul_cycle.append(eul_cycle.pop(0))\n",
    "print eul_cycle\n",
    "\n",
    "#print eul_path\n",
    "fa=open('answer.txt','w')\n",
    "#entry='->'.join(eul_path)\n",
    "#fa.write(entry)\n",
    "fa.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "IOError",
     "evalue": "[Errno 2] No such file or directory: '../../Downloads/dataset_203_5.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-67-34996e55d318>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mre\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m \u001b[0mf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../../Downloads/dataset_203_5.txt'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     34\u001b[0m \u001b[0mlines\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplitlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIOError\u001b[0m: [Errno 2] No such file or directory: '../../Downloads/dataset_203_5.txt'"
     ]
    }
   ],
   "source": [
    "''' FUNCTION '''\n",
    "\n",
    "def EulerianPath(Graph):\n",
    "    from collections import Counter\n",
    "    from collections import defaultdict\n",
    "    from Bioinf_functions import eulerianCycle\n",
    "    # Find ends of graph\n",
    "    outdegrees=Counter(dict([(k,(len(Graph[k]))) for k in Graph]))\n",
    "    indegrees=Counter(value for values in Graph.itervalues() for value in values)\n",
    "    for k in outdegrees+indegrees:\n",
    "        if outdegrees[k]<indegrees[k]:\n",
    "            pre=k\n",
    "        if outdegrees[k]>indegrees[k]:\n",
    "            post=k\n",
    "    # Link ends together to balance graph \n",
    "    if pre:\n",
    "        if pre in Graph:\n",
    "            Graph[pre].append(post)\n",
    "        else:\n",
    "            Graph[pre]=[post]\n",
    "\n",
    "    # run EulerianCycle    \n",
    "    eul_cycle=eulerianCycle(Graph)[1:]\n",
    "    \n",
    "    # Rotate cycle and remove ends link\n",
    "    while eul_cycle[0] != post or eul_cycle[-1] != pre:  \n",
    "        eul_cycle.append(eul_cycle.pop(0))\n",
    "    return eul_cycle\n",
    "\n",
    "''' CHALLENGE '''\n",
    "\n",
    "import re\n",
    "f=open('../../Downloads/dataset_203_5.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "Graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    Graph[data[0]]=data[1:]\n",
    "    \n",
    "fa=open('answer.txt','w')\n",
    "entry='->'.join(EulerianPath(Graph))\n",
    "fa.write(entry)\n",
    "fa.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'dict' object has no attribute 'graph'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-258-6d5cf9012f61>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     37\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m \u001b[0;32mprint\u001b[0m \u001b[0meulerianPath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGraph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-258-6d5cf9012f61>\u001b[0m in \u001b[0;36meulerianPath\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0;31m# find the unbalanced nodes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0mbalances\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdefaultdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0msource\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgraph\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mtarget\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgraph\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m             \u001b[0mbalances\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtarget\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'dict' object has no attribute 'graph'"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "''' ROSALIND CODE \n",
    "cant make it work'''\n",
    "\n",
    "\n",
    "lines=['0 -> 2','1 -> 3','2 -> 1','3 -> 0,4','6 -> 3,7','7 -> 8','8 -> 9','9 -> 6']\n",
    "Graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    Graph[data[0]]=data[1:]\n",
    "    \n",
    "def eulerianPath(self):\n",
    "\n",
    "    # find the unbalanced nodes\n",
    "    balances = defaultdict(int)\n",
    "    for source in self.graph:\n",
    "        for target in self.graph[source]:\n",
    "            balances[target] += 1\n",
    "            balances[source] -= 1\n",
    "            print balances\n",
    "    source, target = None, None\n",
    "    for node in balances:\n",
    "        if balances[node] == -1:\n",
    "            target = node\n",
    "        if balances[node] == 1:\n",
    "            source = node\n",
    "\n",
    "    # add the missing edge\n",
    "    self.graph[source].append(target)\n",
    "\n",
    "    # run eulerianCycle and drop the circle closing postfix\n",
    "    path = self.eulerianCycle()[0:-1]\n",
    "\n",
    "    # rotate until the missing edge matches end and beginning\n",
    "    while path[0] != target or path[-1] != source:  \n",
    "        path.append(path.pop(0))\n",
    "    return path\n",
    "\n",
    "print eulerianPath(Graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAAATGCATCATACGCTCACCCAG\n"
     ]
    }
   ],
   "source": [
    "''' Solve the String Reconstruction Problem.\n",
    "     Input: An integer k followed by a list of k-mers Patterns.\n",
    "     Output: A string Text with k-mer composition equal to Patterns. (If multiple answers exist, you may\n",
    "     return any one.) '''\n",
    "\n",
    "from Bioinf_functions import *\n",
    "\n",
    "def KmerList(word,k):\n",
    "    i=0\n",
    "    liste=[]\n",
    "    while i < len(word)-(k-1):\n",
    "        kmer=word[i:i+k]\n",
    "        liste.append(kmer)\n",
    "        i+=1\n",
    "    liste.sort()\n",
    "    return (liste)\n",
    "\n",
    "def DeBruijn_k(Patterns,k):\n",
    "    from collections import defaultdict\n",
    "    from collections import OrderedDict\n",
    "    \n",
    "    def Prefix(word):\n",
    "        return word[:-1]\n",
    "\n",
    "    def Suffix(word):\n",
    "        return word[1:]\n",
    "    \n",
    "    pref={}\n",
    "    suff={}\n",
    "    link=defaultdict(list)\n",
    "    liste=[]\n",
    "    for word in Patterns:\n",
    "        pref[word]=Prefix(word)\n",
    "        suff[word]=Suffix(word)\n",
    "    for word in Patterns:\n",
    "        if word in sorted(pref):\n",
    "            link[pref[word]].append(suff[word])      \n",
    "    return [str(key)+' -> '+str(','.join(value)) for key,value in sorted(link.items())]\n",
    "\n",
    "\n",
    "def GenomePath2String(path):\n",
    "    def Prefix(word):\n",
    "        return word[:-1]\n",
    "\n",
    "    def Suffix(word):\n",
    "        return word[1:]\n",
    "    seq=path[0]\n",
    "    for n in range(1,len(path)):   \n",
    "        if Prefix(path[n]) ==  Suffix(path[n-1]):\n",
    "                seq+=path[n][-1]\n",
    "    return seq  \n",
    "\n",
    "'''RAW'''\n",
    "\n",
    "#f=open('../../Downloads/rosalind_ba3h.txt','r')\n",
    "#lines=f.read().splitlines()\n",
    "#f.close()\n",
    "lines=[4,'AAAT','AATG','ACCC','ACGC','ATAC','ATCA','ATGC','CAAA','CACC','CATA','CATC','CCAG','CCCA','CGCT','CTCA','GCAT','GCTC','TACG','TCAC','TCAT','TGCA']\n",
    "k=lines[0]\n",
    "Patterns=lines[1:]\n",
    "\n",
    "#print DeBruijn_k(Patterns,k)\n",
    "lines=DeBruijn_k(Patterns,k)\n",
    "Graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    Graph[data[0]]=data[1:]\n",
    "#print Graph\n",
    "#print '->'.join(EulerianPath(Graph))\n",
    "path=EulerianPath(Graph)\n",
    "print GenomePath2String(path)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n",
      "10000001110000001011111111101111110101111101101111100111111100101111011101111010101111001101111000111111000101110110101110101101110100111110100101110011101110010101110001100110001101110000111110000100110000101101101100111101100101101010101101001101101000111101000101100110101100100001100100111100100100101100011101100010001100010101100001101100000001100000111100000100100000101010100111010100101010001101010000111010000100010000101001100111001100101001001101001000111001000101000100111000100101000001101000000000\n"
     ]
    }
   ],
   "source": [
    "''' CODE CHALLENGE: Solve the k-Universal Circular String Problem.\n",
    "     Input: An integer k.\n",
    "     Output: A k-universal circular string.'''\n",
    "from itertools import cycle, islice, dropwhile\n",
    "print 2**4\n",
    "def NumberToPattern(Number, k):\n",
    "    d={'0':0,'1':1}\n",
    "    pattern=[]\n",
    "    if k==1:\n",
    "        #return d.keys()[d.values().index(Number)]\n",
    "        return (list(d.keys())[list(d.values()).index(int(Number))])\n",
    "    prefixIndex = Number//2\n",
    "    r = Number%2\n",
    "    PrefixPattern = NumberToPattern(prefixIndex, (k-1))\n",
    "    symbol = (list(d.keys())[list(d.values()).index(r)])\n",
    "    return (PrefixPattern+symbol)\n",
    "\n",
    "def Overlap(Patterns):\n",
    "    from collections import defaultdict\n",
    "    def Prefix(word):\n",
    "        return word[:-1]\n",
    "\n",
    "    def Suffix(word):\n",
    "        return word[1:]\n",
    "    pref={}\n",
    "    suff={}\n",
    "    liste=[]\n",
    "    for word in Patterns:\n",
    "        pref[Prefix(word)]=word #problem in multiple words have same prefix\n",
    "        suff[Suffix(word)]=word\n",
    "    for word in Patterns:\n",
    "        if Suffix(word) in pref.keys():\n",
    "            liste.append(word+' -> '+pref[Suffix(word)])\n",
    "    return liste                \n",
    "\n",
    "\n",
    "k=9\n",
    "Patterns=[]\n",
    "for number in range(0,2**k):\n",
    "    Patterns.append(NumberToPattern(number, k))\n",
    "\n",
    "#print 'patterns:',Patterns    \n",
    "#print 'overlap',Overlap(Patterns)\n",
    "#print DeBruijn_k(Patterns,k)\n",
    "lines=DeBruijn_k(Patterns,k)\n",
    "#print lines\n",
    "Graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    Graph[data[0]]=data[1:]\n",
    "#print Graph\n",
    "cycle_Leo= eulerianCycle(Graph)\n",
    "'''print cycle_Leo[1:]\n",
    "cycled= cycle(cycle_Leo[1:])  # cycle through cycle_Leo (removing duplicated start position)\n",
    "skipped = dropwhile(lambda x: x != '000', cycled)  # drop the values until x==NewStart\n",
    "sliced = islice(skipped, None, len(cycle_Leo))  # take the first n values\n",
    "cycle_Leo = list(sliced)  # create a list from iterator\n",
    "print cycle_Leo[1:-1]'''\n",
    "\n",
    "print GenomePath2String(cycle_Leo)[k-1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''  StringSpelledByGappedPatterns(GappedPatterns, k, d)\n",
    "        FirstPatterns ← the sequence of initial k-mers from GappedPatterns\n",
    "        SecondPatterns ← the sequence of terminal k-mers from GappedPatterns\n",
    "        PrefixString ← StringSpelledByPatterns(FirstPatterns, k)\n",
    "        SuffixString ← StringSpelledByPatterns(SecondPatterns, k)\n",
    "        for i = k + d + 1 to |PrefixString|\n",
    "            if the i-th symbol in PrefixString does not equal the (i - k - d)-th symbol in SuffixString\n",
    "                return \"there is no string spelled by the gapped patterns\"\n",
    "        return PrefixString concatenated with the last k + d symbols of SuffixString'''\n",
    "\n",
    "\n",
    "''' RAW '''\n",
    "\n",
    "#data\n",
    "k=4\n",
    "d=2\n",
    "\n",
    "Reads_Path=['GACC|GCGC','ACCG|CGCC','CCGA|GCCG','CGAG|CCGG','GAGC|CGGA']\n",
    "GappedPatterns=[]\n",
    "for read in Reads:\n",
    "    GappedPatterns.append(read.split('|'))\n",
    "print GappedPatterns \n",
    "\n",
    "'''f=open('../../Downloads/dataset_6206_7.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "#print lines[0]\n",
    "k=int(lines[0].split(' ')[0])\n",
    "d=int(lines[0].split(' ')[1])\n",
    "#print k,d\n",
    "Reads_Path=lines[1:]\n",
    "GappedPatterns=[]\n",
    "for read in Reads_Path:\n",
    "    GappedPatterns.append(read.split('|'))'''\n",
    "#print len(GappedPatterns) \n",
    "# code\n",
    "#print GappedPatterns[0],GappedPatterns[-1]\n",
    "FirstPatterns=[]\n",
    "for n in range(len(GappedPatterns)): # FirstPatterns long version\n",
    "    FirstPatterns.append(GappedPatterns[n][0])\n",
    "    \n",
    "SecondPatterns=[]\n",
    "for n in range(len(GappedPatterns)):\n",
    "    SecondPatterns.append(GappedPatterns[n][1])    \n",
    "#print FirstPatterns    \n",
    "#FirstPatterns=[reads[n][0] for n in range(len(GappedPatterns))]    \n",
    "#SecondPatterns=[reads[n][1] for n in range(len(GappedPatterns))] \n",
    "PrefixString=GenomePath2String(FirstPatterns)\n",
    "SuffixString=GenomePath2String(SecondPatterns)\n",
    "#print PrefixString,SuffixString\n",
    "for i in range(k + d + 1, len(PrefixString)):\n",
    "    if PrefixString[i] != SuffixString[i - k - d]:\n",
    "        print \"there is no string spelled by the gapped patterns\"\n",
    "print PrefixString,SuffixString,k,d,SuffixString[-(k+d):] \n",
    "\n",
    "r='GACCGAGCGCCGGA'\n",
    "r2='GACCGAGCGCCGGA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGTCTCGCACTACACTCAAGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCAAGTTGCAAGTAGCGCCTTTGTCAGATGAGGGCGCGATGAAGACACCTAGCTTCTCTCAGCACGGCTCAGATTCTCAATTGCCGTGTGTCGAGGGAACAACAGTACAGTTTTACTATTAACATAGATAGACTCTCCTGTCCGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCTAACCCAGTGATGGCTTAGAAGTATGCAGAACTAGGAGGACCCTCATCGACTCTCCTGTCCTCTCTCATCTGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCCCACTCATGCAGTAGGAGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCCGCACGGGCAACACCTGACCTCAAAGCTCTGATCATGGACGAACTGACCGATCATGACTGTACAGCAGGATAGGCGCCCATTGCCGGCTCCTTTCCTACTCGCCCACAGCGACGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCAGTAGGAGCGCACGGGCACTGTGGGGAGTGGTCGCCTGCTTTGCGGCCTGTAGATGCTCTTACCGGTAACAACTACGAGTATGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCACTCTCCTGTCCTCTCTCAGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCTCTCCACTCATGCAGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCTAGGAGCGCACGGGCGCGCCGGTCCGACCAGAGCCAGAAGTCTTGAGTATTTCCAGATCCAGTTTAATACCGTTTAATCAGGCGGGCGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCGGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCATATGGGGTATACGGGTTGTAGTGTCCAGTGGGGTAAAGTTCCATATTATCACTGTGGAAAAAGTTCAGTCTATACATGACGCGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCCAATCAATGGTCTGATATCTCCAATGATTTTGTATAAGTCGGACTGTTTTACAAGTCACGCACGTCGGCATCTGTGAAGATTAAGGATGGCCGCATGAGCATCAGACATGTGAGATCGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCTCGCAGTTCTAGTGCAAGTGTATTTCACAGCCTGTATTTGTGTTTGCCGTCATTATCGGAACTACGCGATTTGCGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCATCCAAATTCACGCGGTCCTGGTTAAATTGCTTCCCTGGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCGGAACTCCACAACCCATCAAGCCTACTAGCTCTTGAAACAAGTTGGTGCTCCGCCCCGTGGTCGGTGAGCTGGAGTCATGACTCTCGACTCTCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCCTGTCCTCTCTCATCTCCACTCATGCAGTAGGAGCGCACGGGCGGATTACATAGGCAACAACGTCCGAGTGTTGGCCCCGGTAAGTCCGGCGCAACTCGATTTAGTCCGTGGTACAAAATGAAAATCGAAAAGACGTTGTCTTCCCAGATCACTGATGTCCCGGGGTCTTGCGATCGGGACCTAGGTGAAATGGGCACACGCGGTTCACT\n"
     ]
    }
   ],
   "source": [
    "''' FUNCTION '''\n",
    "\n",
    "def StringSpelledByGappedPatterns(GappedPatterns, k, d):\n",
    "    ''' description '''\n",
    "    #from Bioinf_functions import GenomePath2String\n",
    "    \n",
    "    FirstPatterns=[]\n",
    "    for n in range(len(GappedPatterns)): # FirstPatterns long version\n",
    "        FirstPatterns.append(GappedPatterns[n][0])\n",
    "    SecondPatterns=[]\n",
    "    for n in range(len(GappedPatterns)):\n",
    "        SecondPatterns.append(GappedPatterns[n][1])\n",
    "        \n",
    "    PrefixString=GenomePath2String(FirstPatterns)\n",
    "    SuffixString=GenomePath2String(SecondPatterns)\n",
    "    for i in range(int(k+d+1), len(PrefixString)):\n",
    "        if PrefixString[i] != SuffixString[i - k - d]:\n",
    "            return \"there is no string spelled by the gapped patterns\"\n",
    "    return PrefixString+SuffixString[-(k+d):] \n",
    "\n",
    "''' Challenge '''\n",
    "\n",
    "f=open('../../Downloads/rosalind_ba3l.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "#print lines[0]\n",
    "k=int(lines[0].split(' ')[0])\n",
    "d=int(lines[0].split(' ')[1])\n",
    "#print k,d\n",
    "Reads_Path=lines[1:]\n",
    "GappedPatterns=[]\n",
    "for read in Reads_Path:\n",
    "    GappedPatterns.append(read.split('|'))\n",
    "#print k,d,len(GappedPatterns) \n",
    "print (StringSpelledByGappedPatterns(GappedPatterns, k, d))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "''' CODE CHALLENGE: Solve the String Reconstruction from Read-Pairs Problem.\n",
    "     Input: Integers k and d followed by a collection of paired k-mers PairedReads.\n",
    "     Output: A string Text with (k, d)-mer composition equal to PairedReads.'''\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "['AC|AT -> CC|TA,CT|TT', 'AT|TG -> TA|GA,TT|GA', 'CA|GA -> AC|AT', 'CC|TA -> CG|AC', 'CG|AC -> GA|CT', 'CT|AG -> TG|GC', 'CT|TT -> TG|TC', 'GA|CT -> AA|TT,AT|TG,AT|TG', 'TA|GA -> AC|AT', 'TC|AA -> CT|AG', 'TG|GC -> GA|CT', 'TG|TC -> GA|CT', 'TT|GA -> TC|AA']\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'pre' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-347-70cc28ec39d1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[0;31m#rint Graph\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     49\u001b[0m \u001b[0;31m#print ' -> '.join(EulerianPath(Graph))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 50\u001b[0;31m \u001b[0mReads_Path\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mEulerianPath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGraph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     51\u001b[0m \u001b[0;31m#print Reads_Path\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m \u001b[0mPairPath\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-67-34996e55d318>\u001b[0m in \u001b[0;36mEulerianPath\u001b[0;34m(Graph)\u001b[0m\n\u001b[1;32m     14\u001b[0m             \u001b[0mpost\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0;31m# Link ends together to balance graph\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mpre\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mpre\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mGraph\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m             \u001b[0mGraph\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpre\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpost\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'pre' referenced before assignment"
     ]
    }
   ],
   "source": [
    "''' RAW '''\n",
    "\n",
    "#f=open('../../Downloads/rosalind_ba3j (1).txt','r')\n",
    "#lines=f.read().splitlines()\n",
    "#f.close()\n",
    "lines=['3 1','ACC|ATA','ACT|ATT','ATA|TGA','ATT|TGA','CAC|GAT','CCG|TAC','CGA|ACT','CTG|AGC','CTG|TTC','GAA|CTT','GAT|CTG','GAT|CTG','TAC|GAT','TCT|AAG','TGA|GCT','TGA|TCT','TTC|GAA']\n",
    "#print lines[0]\n",
    "k=int(lines[0].split(' ')[0])\n",
    "d=int(lines[0].split(' ')[1])\n",
    "print k\n",
    "Reads_Path=lines[1:]\n",
    "GappedPatterns=[]\n",
    "for read in Reads_Path:\n",
    "    GappedPatterns.append(read)\n",
    "#print k,d,GappedPatterns\n",
    "\n",
    "\n",
    "def DeBruijn_k_pairs(GappedPatterns,k):\n",
    "    from collections import defaultdict\n",
    "    from collections import OrderedDict\n",
    "    #from Bioinf_functions import KmerList\n",
    "    \n",
    "    def Prefix_pair(word):\n",
    "        return '|'.join([read[:-1] for read in word.split('|')])\n",
    "\n",
    "    def Suffix_pair(word):\n",
    "        return '|'.join([read[1:] for read in word.split('|')])\n",
    "    \n",
    "    pref={}\n",
    "    suff={}\n",
    "    link=defaultdict(list)\n",
    "    liste=[]\n",
    "    for word in GappedPatterns:\n",
    "        pref[word]=Prefix_pair(word)\n",
    "        suff[word]=Suffix_pair(word)\n",
    "    for word in GappedPatterns:\n",
    "        if word in sorted(pref):\n",
    "            link[pref[word]].append(suff[word])      \n",
    "    return [str(key)+' -> '+str(','.join(value)) for key,value in sorted(link.items())]\n",
    "\n",
    "print DeBruijn_k_pairs(GappedPatterns,k)\n",
    "#lines=DeBruijn_k_pairs(GappedPatterns,k)\n",
    "#print lines\n",
    "Graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    Graph[data[0]]=data[1:]\n",
    "#rint Graph\n",
    "#print ' -> '.join(EulerianPath(Graph))\n",
    "Reads_Path=EulerianPath(Graph)\n",
    "#print Reads_Path\n",
    "PairPath=[]\n",
    "for read in Reads_Path:\n",
    "    PairPath.append(read.split('|'))\n",
    "print PairPath\n",
    "#print StringSpelledByGappedPatterns(PairPath, k, d)\n",
    "FirstPatterns=[]\n",
    "for n in range(len(PairPath)): # FirstPatterns long version\n",
    "    FirstPatterns.append(PairPath[n][0])\n",
    "    print PairPath[n][0]\n",
    "print FirstPatterns\n",
    "\n",
    "SecondPatterns=[]\n",
    "for n in range(len(PairPath)):\n",
    "    SecondPatterns.append(PairPath[n][1])\n",
    "\n",
    "PrefixString=GenomePath2String(FirstPatterns)\n",
    "SuffixString=GenomePath2String(SecondPatterns)\n",
    "\n",
    "print GenomePath2String(FirstPatterns)\n",
    "print GenomePath2String(SecondPatterns)\n",
    "print PrefixString, SuffixString\n",
    "\n",
    "'''for i in range(int(k+d+1), len(PrefixString)):\n",
    "    if PrefixString[i] != SuffixString[i - k - d]:\n",
    "        print \"there is no string spelled by the gapped patterns\"\n",
    "    print PrefixString+SuffixString[-(k+d):]''' \n",
    "PrefixString[k+d+1:]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Graph' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-d587176056ea>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     26\u001b[0m     \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mre\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m' -> |,'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0mgraph\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 28\u001b[0;31m \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mGraph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgraph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmaximalNonBranchingPathes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[0;31m#print result\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Graph' is not defined"
     ]
    }
   ],
   "source": [
    "''' MaximalNonBranchingPaths  '''\n",
    "'''MaximalNonBranchingPaths(Graph)\n",
    "        Paths ← empty list\n",
    "        for each node v in Graph\n",
    "            if v is not a 1-in-1-out node\n",
    "                if out(v) > 0\n",
    "                    for each outgoing edge (v, w) from v\n",
    "                        NonBranchingPath ← the path consisting of the single edge (v, w)\n",
    "                        while w is a 1-in-1-out node\n",
    "                            extend NonBranchingPath by the outgoing edge (w, u) from w \n",
    "                            w ← u\n",
    "                        add NonBranchingPath to the set Paths\n",
    "        for each isolated cycle Cycle in Graph\n",
    "            add Cycle to Paths\n",
    "        return Paths'''\n",
    "\n",
    "''' COULDN'T HACK IT'''\n",
    "''' ROSALIND SOLUTION '''\n",
    "\n",
    "f=open('input.txt','r')\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "#lines=['1 -> 2','2 -> 3','3 -> 4,5','6 -> 7','7 -> 8','8 -> 6', '9 -> 10','10 -> 9']\n",
    "graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    graph[data[0]]=data[1:]\n",
    "result = Graph(graph).maximalNonBranchingPathes()\n",
    "\n",
    "#print result\n",
    "fa=open('output2.txt','w')\n",
    "fa.write('\\n'.join([' -> '.join(path) for path in result]))\n",
    "fa.close()\n",
    "\n",
    "\n",
    "class Graph:\n",
    "    def __init__(self, graph):\n",
    "        self.graph = graph\n",
    "\n",
    "    def countOut(self, node):\n",
    "        return self.getDegrees()[node]['out']\n",
    "\n",
    "    def extendNonBranchingPath(self, first, second, oneInOneOuts):\n",
    "        path, node = [first, second], second\n",
    "        while True:\n",
    "            try:\n",
    "                oneInOneOuts.remove(node)\n",
    "            except:\n",
    "                return path\n",
    "            try:\n",
    "                node = self.graph[node][0]\n",
    "                path.append(node)\n",
    "            except:\n",
    "                return path\n",
    "\n",
    "    def findBranches(self):\n",
    "        branches = defaultdict(list)\n",
    "        for head in self.graph:\n",
    "            if not head in self.getOneInOneOuts():\n",
    "                if self.countOut(head) > 0:\n",
    "                    for tail in self.graph[head]:\n",
    "                        branches[head].append(tail)\n",
    "        return branches\n",
    "    def getDegrees(self):\n",
    "        try:\n",
    "            return self.degrees\n",
    "        except AttributeError:\n",
    "            self.degrees = defaultdict(lambda : defaultdict(int))\n",
    "            for source in self.graph:\n",
    "                for target in self.graph[source]:\n",
    "                    self.degrees[source]['out'] += 1\n",
    "                    self.degrees[target]['in'] += 1\n",
    "            return self.degrees\n",
    "    def getOneInOneOuts(self):\n",
    "        oneInOneOuts = []\n",
    "        degrees = self.getDegrees()\n",
    "        for node in degrees:\n",
    "            if degrees[node]['out'] == 1 and degrees[node]['in'] == 1:\n",
    "                oneInOneOuts.append(node)\n",
    "        return oneInOneOuts\n",
    "\n",
    "    def maximalNonBranchingPathes(self):\n",
    "        pathes = []\n",
    "        oneInOneOuts = self.getOneInOneOuts()\n",
    "        # build non branching pathes\n",
    "        branches = self.findBranches()\n",
    "        for first in branches:\n",
    "            for second in branches[first]:\n",
    "                pathes.append(self.extendNonBranchingPath(first, second,\n",
    "                    oneInOneOuts))\n",
    "        # all remaining nodes are simple cycles\n",
    "        while len(oneInOneOuts) > 0:\n",
    "            path = [oneInOneOuts.pop()]\n",
    "            while True:\n",
    "                node = self.graph[path[-1]][0]\n",
    "                path.append(node)\n",
    "                if path[0] != path[-1]:\n",
    "                    oneInOneOuts.remove(node)\n",
    "                else:\n",
    "                    pathes.append(path)\n",
    "                    break\n",
    "        return pathes\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "''' Contig Generation Problem: Generate the contigs from a collection of reads (with imperfect coverage).\n",
    "     Input: A collection of k-mers Patterns. \n",
    "     Output: All contigs in DeBruijn(Patterns).'''\n",
    "from collections import defaultdict\n",
    "class Graph:\n",
    "    def __init__(self, graph):\n",
    "        self.graph = graph\n",
    "\n",
    "    def countOut(self, node):\n",
    "        return self.getDegrees()[node]['out']\n",
    "\n",
    "    def extendNonBranchingPath(self, first, second, oneInOneOuts):\n",
    "        path, node = [first, second], second\n",
    "        while True:\n",
    "            try:\n",
    "                oneInOneOuts.remove(node)\n",
    "            except:\n",
    "                return path\n",
    "            try:\n",
    "                node = self.graph[node][0]\n",
    "                path.append(node)\n",
    "            except:\n",
    "                return path\n",
    "\n",
    "    def findBranches(self):\n",
    "        branches = defaultdict(list)\n",
    "        for head in self.graph:\n",
    "            if not head in self.getOneInOneOuts():\n",
    "                if self.countOut(head) > 0:\n",
    "                    for tail in self.graph[head]:\n",
    "                        branches[head].append(tail)\n",
    "        return branches\n",
    "    def getDegrees(self):\n",
    "        try:\n",
    "            return self.degrees\n",
    "        except AttributeError:\n",
    "            self.degrees = defaultdict(lambda : defaultdict(int))\n",
    "            for source in self.graph:\n",
    "                for target in self.graph[source]:\n",
    "                    self.degrees[source]['out'] += 1\n",
    "                    self.degrees[target]['in'] += 1\n",
    "            return self.degrees\n",
    "    def getOneInOneOuts(self):\n",
    "        oneInOneOuts = []\n",
    "        degrees = self.getDegrees()\n",
    "        for node in degrees:\n",
    "            if degrees[node]['out'] == 1 and degrees[node]['in'] == 1:\n",
    "                oneInOneOuts.append(node)\n",
    "        return oneInOneOuts\n",
    "\n",
    "    def maximalNonBranchingPathes(self):\n",
    "        pathes = []\n",
    "        oneInOneOuts = self.getOneInOneOuts()\n",
    "        # build non branching pathes\n",
    "        branches = self.findBranches()\n",
    "        for first in branches:\n",
    "            for second in branches[first]:\n",
    "                pathes.append(self.extendNonBranchingPath(first, second,\n",
    "                    oneInOneOuts))\n",
    "        # all remaining nodes are simple cycles\n",
    "        while len(oneInOneOuts) > 0:\n",
    "            path = [oneInOneOuts.pop()]\n",
    "            while True:\n",
    "                node = self.graph[path[-1]][0]\n",
    "                path.append(node)\n",
    "                if path[0] != path[-1]:\n",
    "                    oneInOneOuts.remove(node)\n",
    "                else:\n",
    "                    pathes.append(path)\n",
    "                    break\n",
    "        return pathes\n",
    "    \n",
    "import re\n",
    "#inpt=['ATG','ATG','TGT','TGG','CAT','GGA','GAT','AGA']\n",
    "f=open('../../Downloads/rosalind_ba3k.txt','r')\n",
    "inpt=f.read().splitlines()\n",
    "#print inpt\n",
    "from Bioinf_functions import DeBruijn_k, GenomePath2String\n",
    "\n",
    "lines=DeBruijn_k(inpt,len(inpt[0]))\n",
    "\n",
    "graph={}\n",
    "for line in lines:\n",
    "    data=re.split(' -> |,',line)\n",
    "    graph[data[0]]=data[1:]\n",
    "result = Graph(graph).maximalNonBranchingPathes()\n",
    "\n",
    "entry= '\\n'.join([GenomePath2String(path) for path in result])\n",
    "#print entry\n",
    "\n",
    "fa=open('output.txt','w')\n",
    "fa.write(entry)\n",
    "fa.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUGGCCAUGGCCCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA UCACCCGUUAAUACGGGUACUAUUGAUCUCAGUUCUGGGGGCCAUGGCCAU\n",
      "[['AUG'], ['GCA', 'GCG', 'GCC', 'GCU']]\n",
      "['AUGGCC', 'AUGGCC', 'ACGGGU']\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
