{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "262144\n",
      "3.81469726562e-06\n",
      "992\n",
      "496000\n",
      "1.89208984375\n"
     ]
    }
   ],
   "source": [
    "from __future__ import division\n",
    "#How many 9-mers exists? (this will not enter in the calculus)\n",
    "a=(4**9)\n",
    "print a\n",
    "#Whats the probability of, in a set of the 9-mers of Hint 1, you pick a 9-mer that you want in the first pick?\n",
    "b=1/a\n",
    "print b\n",
    "#How many 9-mers exists in a string with 1000 nucleotides?\n",
    "c=1000-(9-1)\n",
    "print c\n",
    "#How many 9-mers exists in a set of 500 strings on length 1000?\n",
    "d= 500*c\n",
    "print d\n",
    "#What is the expected number of occurrences of the 9-mer that you choose in Hint 2 occur in the set of Hint 4?\n",
    "e=b*d\n",
    "print e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.89208984375\n"
     ]
    }
   ],
   "source": [
    "'''EXERCISE BREAK: What is the expected number of occurrences of a 9-mer in 500 random DNA strings,\n",
    "each of length 1000? Assume that the sequences are formed by selecting each nucleotide (A, C, G, T)\n",
    "with the same probability (0.25).'''\n",
    "\n",
    "k=9\n",
    "len_string=1000\n",
    "no_strings=500\n",
    "\n",
    "p=(1/4**k)*no_strings*(len_string-k+1)\n",
    "print p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCGTT GCCTA GCCCG GCCCA GTAGA TTCTA CTAAG CCTAT CTAAA GCACT GGGTA GGGTC GCGTA GCGTC CTAAT CCTAA GCACA CCTAC CGAGC AGATA ACGTT CCTAG AGGCC AGGCG GGCTA GGCTC TAAGC AGACC CATGG AGACG CATGC GCGCT AGGCT TTCGA TTACA CGCAT TCGCA CCAAT CGAAC GCTGA CCCTA TATCC AACGG CGAAT ACCTA GTTCC AGCAT GTTCA CAGCG CAAAT CAGCC CTGGA TCATG CTGGC TCATA GTCGA CTTGG GGTCC GAGTA AACAC GGTCG ACCGA ACCGG ACTGG ACGGG TGAAC ACTGC ATCGC ACCGT GACTC TATAC GACTA GCCAA GGTCA GGCCT ACCGC CTAGA TGTAC CGACA CCGAT CACTT CAAAC AAGTC GGCCG CCGTA AACGC CCTTG CAGAT CAAGC AGTCG TAATC TGGCC AGTCC CATAC GCGAA GGACG GGACC GCGGC GATCC CGTCG CTCGA CGACC GCGCA CTGAT ATGCA CGACG GCTTA CACAT AGCCG ATGCC AAGGC CCACT CTGAG TTCCA AATGC GAAGC CTGAA CCACG CAGTG ACCAG AGCCT AACGA CCACA ATCCT ACTCC GAGGC CTCAT GACGC ACGGT AACGT ATACC GACGG ACGAT CGTCA ACTAC GGGCG TCCAG CTCAG CTATC ATCCA TTGCG TGCCC ACGAC ACTCG GGAAC ACGAG TGACC CTATG TGCAC ACGCC GCATT GCCGA GTACC CTTAT GTCCG GTACG TCCTA GTCCC GCATC ACGCT GCAGA TCAAC GTCCT AGGTC TACTG TGGAC GCGGA AGGTG TACTC TACCG TAACG CATCT TAACC TAACA GTGCA ACGCG AGCTC CGATT AGCGT CATCC AGGGC CTTAC CCTGG AGCGC CCCGA AGCTT CGATA AGCGG CGATG ACATA ACACG ACATC CTTAA GCTCC CACGA GCTCA CTCGC CCATT CATCG ACACT CCCAT CGATC CTGCG CTTCG CCATC CCATA CGTAT GACAT TCCGT GGGAT GAATG TACGA GAATA GAATC ATGCT CGTAG CGTAA TCCGG GAATT CTCCA TCGAA CGGTC ACTTA TAGCA ACTTC TAGCC AAACG ACGTG GGATT CGTAC AAACC ACGTA CAACT TATGC CTACT CGGAT ATCTA GGATA TCGTA CAACG ATACG GCAAA CAATG TGGGC CTGAC CAATC TGAAT AGGAC ATCCG TACGC CCGAG CGCCG AGTGC TCTAG TAAAC GGTAC TCGAC GGAGT GCATA GATGC ATCCC AAGCG CGAGT ATGGC GAGCT CGCTA CCAGA GCTAA GCTAC CGAGG AAGCT ATTCA ATTCC AACTC AATCC GAACC TCGGA GAACG CGCGT TTCGG CCCTG CGACT TTCGC AACCG ACCTC ATCGG TATTC TGTCC CGTGC GCATG TGCGA GTCTA ACTCA GACCG GACCC TCTAA GACCA\n"
     ]
    }
   ],
   "source": [
    "###load functions\n",
    "def HammingDistance(p,q):\n",
    "    HamDi=int(0)\n",
    "    for i in range(len(p)):\n",
    "        if p[i]==q[i]:\n",
    "            continue\n",
    "        else:\n",
    "            HamDi+=+1\n",
    "    return(HamDi)\n",
    "#--\n",
    "def Neighbors(Pattern, d):\n",
    "    if d == 0:\n",
    "        Neighborhood =(str(Pattern))\n",
    "        return Neighborhood\n",
    "    if len(Pattern) == 1: \n",
    "        Neighborhood=['A','C','G','T']\n",
    "        return Neighborhood\n",
    "    nucl=['A','C','G','T']\n",
    "    Neighborhood=list()\n",
    "    SuffixNeighbors = Neighbors((Pattern[1:]), d)\n",
    "    for Text in SuffixNeighbors:\n",
    "        if HammingDistance((Pattern[1:]), Text) < d:\n",
    "            for n in nucl:\n",
    "                Neighborhood.append(str(n + Text)) \n",
    "        else:\n",
    "            Neighborhood.append(str(Pattern[0]+ Text))\n",
    "    return Neighborhood\n",
    "#-----------------------------------------------------------\n",
    "#f = open('/Users/juniper_admin/Desktop/dataset_156_7.txt')\n",
    "f = open('../../Downloads/rosalind_ba2a.txt')\n",
    "\n",
    "\n",
    "lines = [item.strip() for item in f.readlines()]\n",
    "k,d = map(int,lines[0].split(' '))\n",
    "Dna=lines[1:]\n",
    "f.close()\n",
    "#--\n",
    "def MOTIFENUMERATION(Dna, k, d):\n",
    "    i=1\n",
    "    p=list()\n",
    "    while i<len(lines):\n",
    "        #print lines[i]\n",
    "        list_i=list()\n",
    "        for n in range(len(lines[i])-k+1):\n",
    "            Neigh=Neighbors(lines[i][n:n+k],d)\n",
    "            for word in Neigh:\n",
    "                list_i.append(word)\n",
    "        #print (list_i) \n",
    "        p.append(list_i)\n",
    "        i=i+1\n",
    "    #print p    \n",
    "    return(set.intersection(*map(set, p)))\n",
    "\n",
    "print (' '.join(MOTIFENUMERATION(Dna, k, d)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9.91629000536\n"
     ]
    }
   ],
   "source": [
    "'''13.7162900054'''\n",
    "liste=[0.2,0.7,0.1,0.2,0.2,0.6,1,1,0.9,0.1,0.9,0.1,0.9,0.1,0.1,0.4,0.5,0.8,0.1,0.1,0.2,0.7,0.1,0.4,0.3,0.3,0.6,0.4]\n",
    "c_it=0\n",
    "for item in liste:\n",
    "    c_it+=-(item*m.log(item,2))\n",
    "print c_it    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(['AAAAGCACAGATGTCTAAACTTCGGCCGCCGACGCCAGGGGC', 'GGGTTAGACTAGCTCTCTTGTTCTGTCCGAGAGCGTGCGGCC', 'GCCGCCAATCCCTAATTCGTCATTATCATCTGAGTGTCAGTG', 'GCAGCCGAACGAGATCGTCTAATGGCTGCTTCCTGTTGGGGG', 'ATGGGAGCAGCCATGCACATAACTTTCTGTGTAACTTGGACG', 'TCACGTATGGGAGCAGCCCACACTCGTGTTTACCCTATGGCA', 'TGGTTATCATATGCTGCCAGCTAGCACACCACTCGGGGATGT', 'TTCTAGCGTCGTTAACTGCAATTACCCTACCCAACTGCAGCC', 'CGAGCTGCGTATGCGGCCGCAAGCACCTAGTATTTGCTCGAA', 'AGGATGACGAAAGTCACGTCTTGTATCAACAACGGCGCCGCC'], 6)\n",
      "GCAGCC\n",
      "The script took 5.40172195435 s\n"
     ]
    }
   ],
   "source": [
    "'''    MEDIANSTRING(Dna, k)\n",
    "        distance ← ∞\n",
    "        for each k-mer Pattern from AA…AA to TT…TT\n",
    "            if distance > d(Pattern, Dna)\n",
    "                 distance ← d(Pattern, Dna)\n",
    "                 Median ← Pattern\n",
    "        return Median'''\n",
    "#--------------------------------------------------------------------------------------\n",
    "def NumberToPattern(Number, k):\n",
    "    d={'A':0,'C':1,'G':2,'T':3}\n",
    "    pattern=[]\n",
    "    if k==1:\n",
    "        #return d.keys()[d.values().index(Number)]\n",
    "        return (list(d.keys())[list(d.values()).index(int(Number))])\n",
    "    prefixIndex = Number//4\n",
    "    r = Number%4\n",
    "    PrefixPattern = NumberToPattern(prefixIndex, (k-1))\n",
    "    symbol = (list(d.keys())[list(d.values()).index(r)])\n",
    "    return (PrefixPattern+symbol)\n",
    "\n",
    "#--\n",
    "def d(Pattern, Dna):\n",
    "    score=0\n",
    "    k=len(Pattern)\n",
    "    for i in range(0,len(Dna)):\n",
    "        Hamdist={}\n",
    "        minDist={}\n",
    "        seq=Dna[i]\n",
    "        for n in range(0,len(seq)-(k-1)):\n",
    "            Hamdist[seq[n:n+k]]=HammingDistance(Pattern,seq[n:n+k])\n",
    "        minimum=int(min(Hamdist.values()))\n",
    "        score+=minimum\n",
    "    return(score)\n",
    "\n",
    "\n",
    "def MEDIANSTRING(Dna, k):\n",
    "    distance= float('inf')\n",
    "    kmer_pattern=list()\n",
    "    for n in range(4**k):\n",
    "        kmer_pattern.append(NumberToPattern(n, k))\n",
    "\n",
    "    for Pattern in kmer_pattern:\n",
    "        if distance > d(Pattern, Dna):\n",
    "            distance = d(Pattern, Dna)\n",
    "            Median=Pattern\n",
    "    return(Median)\n",
    "##################################################\n",
    "##################################################\n",
    "\n",
    "#f = open('/Users/juniper_admin/Desktop/dataset_158_9.txt','r')\n",
    "f = open('../../Downloads/rosalind_ba2b.txt')\n",
    "\n",
    "lines = f.read().splitlines()\n",
    "f.close()\n",
    "k=int(lines[0])\n",
    "Dna=lines[1:]\n",
    "#print(Dna,k)\n",
    "\n",
    "#--\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "print(MEDIANSTRING(Dna, k))\n",
    "print ('The script took '+str(time.time()-start_time)+' s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCATTAT\n"
     ]
    }
   ],
   "source": [
    "'''Profile-most Probable k-mer Problem'''\n",
    "\n",
    "from numpy import *\n",
    " \n",
    "#--------------------------------------------------------\n",
    "\n",
    "def PatternToNumber(Pattern):\n",
    "    d={'A':0,'C':1,'G':2,'T':3}\n",
    "    if Pattern == '':\n",
    "        return 0\n",
    "    symbol = Pattern[-1]\n",
    "    Pattern= Pattern[:-1]\n",
    "    return(4 * PatternToNumber(Pattern) + d[symbol])\n",
    "\n",
    "#--\n",
    "\n",
    "\n",
    "def Profile_most(Text,k,Profile):\n",
    "    max_score=0\n",
    "    most_probable_kmer=Text[0:0+k]\n",
    "    for n in range(0,len(Text)-(k-1)):\n",
    "        kmer=Text[n:n+k]\n",
    "        score=1\n",
    "        for i in range(0,len(kmer)):\n",
    "            probaN=Profile[PatternToNumber(kmer[i])][i]\n",
    "            score *= probaN\n",
    "        if score>max_score:\n",
    "            max_score=score\n",
    "            #global most_probable_kmer\n",
    "            most_probable_kmer=kmer\n",
    "        else:\n",
    "            continue\n",
    "    return(most_probable_kmer)\n",
    "\n",
    "#-------------------------------------------------------\n",
    "#f=open(\"/Users/juniper_admin/Dropbox/Maeva+Connor/dataset_4_4.txt\")\n",
    "f = open('../../Downloads/rosalind_ba2c.txt')\n",
    "\n",
    "lines=f.read().splitlines()\n",
    "Text=lines[0]\n",
    "k=int(lines[1])\n",
    "matrix=loadtxt(lines[2:])\n",
    "Profile=matrix\n",
    "\n",
    "#print (Text,k,Profile)\n",
    "print((Profile_most(Text,k,Profile)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "global name 'array' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-4a2771b5a967>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mDna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlines\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m \u001b[0;32mprint\u001b[0m \u001b[0mGREEDYMOTIFSEARCH\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDna\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/juniper_admin/Dropbox/BIOINF/Bioinf_functions.pyc\u001b[0m in \u001b[0;36mGREEDYMOTIFSEARCH\u001b[0;34m(Dna, k, t)\u001b[0m\n\u001b[1;32m    725\u001b[0m         \u001b[0mMotif_list\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    726\u001b[0m         \u001b[0mMotif_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmotif\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 727\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDna\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    728\u001b[0m             \u001b[0mText\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mDna\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    729\u001b[0m             \u001b[0mProfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mMakeProfile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mMotif_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: global name 'array' is not defined"
     ]
    }
   ],
   "source": [
    "'''Greedy motif search with pseudocount'''\n",
    "\n",
    "from numpy import *\n",
    "import numpy as np\n",
    "import time\n",
    "start_time = time.time()\n",
    "#--\n",
    "def HammingDistance(p,q):\n",
    "    HamDi=int(0)\n",
    "    for i in range(len(p)):\n",
    "        if p[i]==q[i]:\n",
    "            continue\n",
    "        else:\n",
    "            HamDi+=+1\n",
    "    return(HamDi)\n",
    "\n",
    "#--\n",
    "def NumberToPattern(Number, k):\n",
    "    d={'A':0,'C':1,'G':2,'T':3}\n",
    "    pattern=[]\n",
    "    if k==1:\n",
    "        #return d.keys()[d.values().index(Number)]\n",
    "        return (list(d.keys())[list(d.values()).index(int(Number))])\n",
    "    prefixIndex = Number//4\n",
    "    r = Number%4\n",
    "    PrefixPattern = NumberToPattern(prefixIndex, (k-1))\n",
    "    symbol = (list(d.keys())[list(d.values()).index(r)])\n",
    "    return (PrefixPattern+symbol)\n",
    "#--\n",
    "def PatternToNumber(Pattern):\n",
    "    d={'A':0,'C':1,'G':2,'T':3}\n",
    "    if Pattern == '':\n",
    "        return 0\n",
    "    symbol = Pattern[-1]\n",
    "    Pattern= Pattern[:-1]\n",
    "    return(4 * PatternToNumber(Pattern) + d[symbol])\n",
    "#--\n",
    "\n",
    "def MakeProfile(Motif_list):\n",
    "    k=len(Motif_list[0])\n",
    "    matrix_occ = np.zeros((4,k))\n",
    "    matrix_occ +=1\n",
    "    for i in range(len(Motif_list)):\n",
    "        for n in range(len(Motif_list[i])):\n",
    "            matrix_occ[(PatternToNumber(Motif_list[i][n])),n]+=1   \n",
    "    Profile_matrix= matrix_occ / matrix_occ.sum(axis=0, dtype='float')\n",
    "    #return matrix_occ.sum(axis=0, dtype='float')\n",
    "    return Profile_matrix  \n",
    "#--\n",
    "def Profile_most(Text,k,Profile):\n",
    "    max_score=0\n",
    "    most_probable_kmer=Text[0:0+k]\n",
    "    for n in range(0,len(Text)-(k-1)):\n",
    "        kmer=Text[n:n+k]\n",
    "        score=1\n",
    "        for i in range(0,len(kmer)):\n",
    "            probaN=Profile[PatternToNumber(kmer[i])][i]\n",
    "            score *= probaN\n",
    "        if score>max_score:\n",
    "            max_score=score\n",
    "            most_probable_kmer=kmer\n",
    "        else:\n",
    "            continue\n",
    "    return(most_probable_kmer)\n",
    "#--\n",
    "def Consensus(Profile):\n",
    "    maxarg= list(Profile.argmax(0))\n",
    "    Cons=''\n",
    "    for posi in maxarg:\n",
    "        Cons+=NumberToPattern(int(posi), 1)\n",
    "    return Cons\n",
    "#--\n",
    "def d(Pattern, Dna):\n",
    "    score=0\n",
    "    k=len(Pattern)\n",
    "    for i in range(0,len(Dna)):\n",
    "        Hamdist={}\n",
    "        minDist={}\n",
    "        seq=Dna[i]\n",
    "        for n in range(0,len(seq)-(k-1)):\n",
    "            Hamdist[seq[n:n+k]]=HammingDistance(Pattern,seq[n:n+k])\n",
    "        minimum=int(min(Hamdist.values()))\n",
    "        score+=minimum\n",
    "    return(score)\n",
    "#---------------------------------------------------------------------\n",
    "#f = open('/Users/juniper_admin/Desktop/motifen.txt','r')\n",
    "f = open('../../Downloads/rosalind_ba2e.txt')\n",
    "\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "k=int(lines[0].split(\" \")[0])\n",
    "t=int(lines[0].split(\" \")[1])\n",
    "Dna=lines[1:]\n",
    "\n",
    "\n",
    "#starting conditions:\n",
    "BestMotifs_list=list()\n",
    "for string in range(len(Dna)):\n",
    "    BestMotifs_list.append(Dna[string][0:0+k])\n",
    "\n",
    "#list all kmers in first string\n",
    "kmer_1st_string=list()\n",
    "for n in range(len(Dna[0])-(k-1)):\n",
    "    Motif=Dna[0][n:n+k]\n",
    "    kmer_1st_string.append(Motif)\n",
    "#-     \n",
    "for motif in kmer_1st_string:\n",
    "    \n",
    "    Motif_list=list()    \n",
    "    Motif_list.append(motif)\n",
    "    for i in range(1,len(Dna)):\n",
    "        Text=Dna[i]\n",
    "        Profile=array(MakeProfile(Motif_list))\n",
    "        new_motif=Profile_most(Text,k,Profile) \n",
    "        Motif_list.append(new_motif)\n",
    "    Motif_score= d(Consensus(MakeProfile(Motif_list)), Motif_list)\n",
    "    if Motif_score < d(Consensus(MakeProfile(BestMotifs_list)), BestMotifs_list):\n",
    "        BestMotifs_list=Motif_list\n",
    "print('\\n'.join(BestMotifs_list))\n",
    "#print ('The script took '+str(time.time()-start_time)+' s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGGTTCATCGGA\n",
      "TAGAAGGATTCA\n",
      "GTCTTATAGAGC\n",
      "AGGTACTACGCC\n",
      "ATCAAGAACTGC\n",
      "AGGATCTTGAGA\n",
      "ATGTTTCGAACT\n",
      "GTGCTGGGGTTC\n",
      "GTCTTCTGTAGC\n",
      "AGGTTCTGCAGA\n",
      "GTCTTGTGCAGC\n",
      "GGCTTATTCAGC\n",
      "GTCTTCTGAAGC\n",
      "GGCTTTTGGAGC\n",
      "AGCATGCGTACA\n",
      "GTCTTCTGGAGC\n",
      "GGCTTCTTAAGC\n",
      "GGCTTATAAAGC\n",
      "AGCTACTGTACT\n",
      "GGCCTGGGTGCC\n",
      "GTCTATTGCGGT\n",
      "AGGCACGGAAGA\n",
      "GGCTTGTACAGC\n",
      "GGGTTAGGAAGA\n",
      "GACTTATTAAGC\n"
     ]
    }
   ],
   "source": [
    "'''Greedy motif search without pseudocount'''\n",
    "\n",
    "from numpy import *\n",
    "import numpy as np\n",
    "import time\n",
    "start_time = time.time()\n",
    "#--\n",
    "def HammingDistance(p,q):\n",
    "    HamDi=int(0)\n",
    "    for i in range(len(p)):\n",
    "        if p[i]==q[i]:\n",
    "            continue\n",
    "        else:\n",
    "            HamDi+=+1\n",
    "    return(HamDi)\n",
    "\n",
    "#--\n",
    "def NumberToPattern(Number, k):\n",
    "    d={'A':0,'C':1,'G':2,'T':3}\n",
    "    pattern=[]\n",
    "    if k==1:\n",
    "        #return d.keys()[d.values().index(Number)]\n",
    "        return (list(d.keys())[list(d.values()).index(int(Number))])\n",
    "    prefixIndex = Number//4\n",
    "    r = Number%4\n",
    "    PrefixPattern = NumberToPattern(prefixIndex, (k-1))\n",
    "    symbol = (list(d.keys())[list(d.values()).index(r)])\n",
    "    return (PrefixPattern+symbol)\n",
    "#--\n",
    "def PatternToNumber(Pattern):\n",
    "    d={'A':0,'C':1,'G':2,'T':3}\n",
    "    if Pattern == '':\n",
    "        return 0\n",
    "    symbol = Pattern[-1]\n",
    "    Pattern= Pattern[:-1]\n",
    "    return(4 * PatternToNumber(Pattern) + d[symbol])\n",
    "#--\n",
    "def MakeProfile(Motif_list):\n",
    "    k=len(Motif_list[0])\n",
    "    matrix_occ = np.zeros((4,k))\n",
    "    matrix_occ +=1\n",
    "    for i in range(len(Motif_list)):\n",
    "        for n in range(len(Motif_list[i])):\n",
    "            matrix_occ[(PatternToNumber(Motif_list[i][n])),n]+=1   \n",
    "    Profile_matrix= matrix_occ / matrix_occ.sum(axis=0, dtype='float')\n",
    "    #return matrix_occ.sum(axis=0, dtype='float')\n",
    "    return Profile_matrix  \n",
    "\n",
    "def MakeProfile(Motif_list):\n",
    "    '''without pseudocount'''\n",
    "    k=len(Motif_list[0])\n",
    "    matrix_occ = np.zeros((4,k))\n",
    "    for i in range(len(Motif_list)):\n",
    "        for n in range(len(Motif_list[i])):\n",
    "            matrix_occ[(PatternToNumber(Motif_list[i][n])),n]+=1   \n",
    "    Profile_matrix= matrix_occ / matrix_occ.sum(axis=0, dtype='float')       \n",
    "    return Profile_matrix  \n",
    "#--\n",
    "def Profile_most(Text,k,Profile):\n",
    "    max_score=0\n",
    "    most_probable_kmer=Text[0:0+k]\n",
    "    for n in range(0,len(Text)-(k-1)):\n",
    "        kmer=Text[n:n+k]\n",
    "        score=1\n",
    "        for i in range(0,len(kmer)):\n",
    "            probaN=Profile[PatternToNumber(kmer[i])][i]\n",
    "            score *= probaN\n",
    "        if score>max_score:\n",
    "            max_score=score\n",
    "            most_probable_kmer=kmer\n",
    "        else:\n",
    "            continue\n",
    "    return(most_probable_kmer)\n",
    "#--\n",
    "def Consensus(Profile):\n",
    "    maxarg= list(Profile.argmax(0))\n",
    "    Cons=''\n",
    "    for posi in maxarg:\n",
    "        Cons+=NumberToPattern(int(posi), 1)\n",
    "    return Cons\n",
    "#--\n",
    "def d(Pattern, Dna):\n",
    "    score=0\n",
    "    k=len(Pattern)\n",
    "    for i in range(0,len(Dna)):\n",
    "        Hamdist={}\n",
    "        minDist={}\n",
    "        seq=Dna[i]\n",
    "        for n in range(0,len(seq)-(k-1)):\n",
    "            Hamdist[seq[n:n+k]]=HammingDistance(Pattern,seq[n:n+k])\n",
    "        minimum=int(min(Hamdist.values()))\n",
    "        score+=minimum\n",
    "    return(score)\n",
    "#---------------------------------------------------------------------\n",
    "#f = open('/Users/juniper_admin/Desktop/motifen.txt','r')\n",
    "f = open('../../Downloads/rosalind_ba2d.txt')\n",
    "\n",
    "lines=f.read().splitlines()\n",
    "f.close()\n",
    "\n",
    "k=int(lines[0].split(\" \")[0])\n",
    "t=int(lines[0].split(\" \")[1])\n",
    "Dna=lines[1:]\n",
    "\n",
    "\n",
    "#starting conditions:\n",
    "BestMotifs_list=list()\n",
    "for string in range(len(Dna)):\n",
    "    BestMotifs_list.append(Dna[string][0:0+k])\n",
    "\n",
    "#list all kmers in first string\n",
    "kmer_1st_string=list()\n",
    "for n in range(len(Dna[0])-(k-1)):\n",
    "    Motif=Dna[0][n:n+k]\n",
    "    kmer_1st_string.append(Motif)\n",
    "#-     \n",
    "for motif in kmer_1st_string:\n",
    "    \n",
    "    Motif_list=list()    \n",
    "    Motif_list.append(motif)\n",
    "    for i in range(1,len(Dna)):\n",
    "        Text=Dna[i]\n",
    "        Profile=array(MakeProfile(Motif_list))\n",
    "        new_motif=Profile_most(Text,k,Profile) \n",
    "        Motif_list.append(new_motif)\n",
    "    Motif_score= d(Consensus(MakeProfile(Motif_list)), Motif_list)\n",
    "    if Motif_score < d(Consensus(MakeProfile(BestMotifs_list)), BestMotifs_list):\n",
    "        BestMotifs_list=Motif_list\n",
    "print('\\n'.join(BestMotifs_list))\n",
    "#print ('The script took '+str(time.time()-start_time)+' s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def Profile_most(Text,k,Profile):\n",
    "    max_score=0\n",
    "    most_probable_kmer=Text[0:0+k]\n",
    "    for n in range(0,len(Text)-(k-1)):\n",
    "        kmer=Text[n:n+k]\n",
    "        score=1\n",
    "        for i in range(0,len(kmer)):\n",
    "            probaN=Profile[PatternToNumber(kmer[i])][i]\n",
    "            return probaN\n",
    "            score *= probaN\n",
    "            return score\n",
    "        if score>max_score:\n",
    "            max_score=score\n",
    "            most_probable_kmer=kmer\n",
    "\n",
    "    return(most_probable_kmer)\n",
    "\n",
    "def Pr(kmer,Profile):\n",
    "    score=1\n",
    "    for i in range(0,len(kmer)):\n",
    "        probaN=Profile[PatternToNumber(kmer[i])][i]\n",
    "        score *= probaN\n",
    "    return score\n",
    "\n",
    "Random\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
